# Overview

## Functional analysis of -Omics data 
Workshop 2024

### General information

The workshop covers the bioinformatics concepts and tools available for interpreting a gene list using gene ontology and pathway information. The workshop focuses on the principles and concepts required for analyzing and conducting functional and pathway analysis on a gene list from any organism, although the focus will be on human and model eukaryotic organisms.

### Course Objectives

Participants will gain practical experience and skills to be able to:

- Understand basic concepts of functional enrichment analysis;
- Interpret enrichment analysis results;
- Get systems perspective of gene functions;
- Get more information about a gene list;
- Discover what pathways are enriched in a gene list (and use it for hypothesis generation);
- Predict gene function and extend a gene list;
- Follow workflow after the workshop to conduct their own analysis.

### Target Audience

This workshop is intended for biologists working with ‘-Omics data’ (e.g. RNA-Seq, protein expression and other omics data), who are interested in interpreting large gene/protein lists resulting from their experiments.

### Setup Requirements

This workshop will be delivered online over zoom; you may wish to install the dedicated zoom. Otherwise, no special software installation will required, as we will be using online analysis tools.  

 * Zoom Link: 


Links and material will be provided on the day. BYO coffee.


### Schedule

```{r, echo=FALSE, message=FALSE}
library(knitr)
suppressWarnings(library(kableExtra))
library(dplyr)

# Create the table data
schedule <- data.frame(
  Day = c("Day 1", "", "", "", "", "", "", "", "", "", "", "Day 1", "Day 2", "", "", "", "", "", "", "", "", "Day 2"),
  Instructor = c("", "HK", "HK", "HK", "HK", "HK", "HK", "HK", "HK", "HK", "HK", "", "", "HK", "CW", "CW", "CW", "CW", "CW", "CW", "CW", ""),
  Activity = c("Welcome and housekeeping", 
               "Introduction", 
               "Data acquisition", 
               "Filtering gene list", 
               "Hands-on with Interactive Calculator (breakout rooms); https://bioinformatics3.erc.monash.edu/rsconnect/content/241/", 
               "gProfiler [GO + pathways] (https://biit.cs.ut.ee/gprofiler/gost)", 
               "Hands-on with gProfiler (breakout rooms)", 
               "Break", 
               "STRING (https://string-db.org/)", 
               "Reactome (https://reactome.org/)", 
               "GSEA (GenePattern) (https://cloud.genepattern.org/gp/pages/index.jsf)", 
               "3 hrs", 
               "Welcome and housekeeping", 
               "Day -1 recap", 
               "Using R for functional enrichment analysis; Applications and advantages; Working with confidential data; Customisation, flexibility, reproducibility; Automation and batch processing", 
               "Available packages in R -; Clusterprofiler; Gprofiler; Any other?", 
               "Introducing R, R Markdown, Rstudio; Getting logged on RStudio environment; Discuss R Markdown; Discuss basic features of Rstudio", 
               "Clusterprofiler - Handon; Breakout rooms; Work on and discuss results based on following criterion; Analysis; ORA; GSEA; Ontologies; GO; Pathway (KEGG, Reactome); …; Visualisations", 
               "gprofiler - Handson; Breakout rooms; Work on and discuss specific features; gost function with standard analysis and plots - Discuss how the plots from gprofiler are different (than clusterprofiler) and also useful; Send  analysis from R to g:Profiler web interface ; Sharing the results easily with colleagues ; To accompany a publication without the peers having to run the full analysis code in R; Integrating results with external tools for visualisations; Alter results using ggplot2, enrichplot, clusterProfiler; Using custom annotations; Non-model organisms, that are not annotated in the Ensembl database; Enable users to upload custom annotation files", 
               "Experiment wrap up  ; Discuss results; Enrichments look different from different tools - Why", 
               "Wrap up and feedback", 
               "3 hrs"),
  `Time (mins)` = c(10, 10, 5, 15, 15, 20, 20, 15, 20, 20, 30, "", 5, 15, 30, 5, 30, 30, 30, 30, 5, ""),
  check.names = FALSE 
)

# Render the table with specified column widths
kable(schedule, "html", escape = FALSE) %>%
  kable_styling(full_width = FALSE, position = "center") %>%
  column_spec(1, width = "10em", bold = TRUE) %>%  # Adjust width
  column_spec(3, width = "25em")  # Center align
```


<!-- ### Abbreviations -->

- **HK**: Hossein V Kahrood
- **CW**: Cali Willet
