# Online Tools

Functional enrichment analysis can be performed using various web-based tools, each of which is designed to meet specific analytical needs. These tools often vary in the databases they use, their statistical approaches, and their capabilities to perform different types of analysis, such as Over-Representation Analysis (ORA) or Gene Set Enrichment Analysis (GSEA).

In this workshop, we will explore several popular tools for functional enrichment analysis, including gProfiler, STRING, Reactome, and MSigDB GSEA. Each tool offers unique features and insights, providing flexibility in selecting the right method for diverse datasets and research questions.


## FEA in [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost)
gProfiler is known for its integration of numerous species and databases. It supports both ORA and GSEA, enabling users to assess Gene Ontology (GO), biological pathways, regulatory motifs and protein databases. With gProfile one can 

### Steps to perform ORA in g:Profiler:

<span style="color:orange;">- Prepare Input List:</span> Ensure your input is formatted as one gene per line or in a suitable format for g:Profiler.

<span style="color:orange;">- Input Gene List:</span> Paste your prepared gene list directly into the input box on the g:Profiler web page or upload a file containing your list.

<span style="color:orange;">- Select Organism:</span> Choose the appropriate organism from the `Organism` dropdown menu (e.g., *Homo sapiens* for human data).

<span style="color:orange;">- Choose Statistical Domain Scope:</span> Under `Advanced options`, select your preferred statistical background from the `Statistical domain scope` menu.If you choose "Custom" background, provide your custom background list by pasting or uploading the relevant file.

<span style="color:orange;">- Set Significance Threshold:</span> Select the desired significance threshold method, such as *g:SCS*, *Bonferroni*, or *Benjamini-Hochberg*.
   - Specify the threshold value (e.g., 0.05, 0.1, etc.).

<span style="color:orange;">- Select Functional Annotation Databases: </span> Navigate to the `Data sources` tab and choose one or more databases for analysis. Available options include:

  - *Gene Ontology (GO)*: Biological Process, Molecular Function, and Cellular Component.
  - *KEGG Pathways*
  - *Reactome Pathways*
  - *WikiPathways*
  - *TRANSFAC*
  - *mirTarBase*
  - *Human Protein Atlas*
  - *CORUM*
  - *Human Phenotype Ontology (HP)*

<span style="color:orange;">- Run Query:</span> Run the analysis and review the enriched terms, pathways, and visual outputs. Download the results as needed for further exploration.

#### Browse the Results

- **Overview**:
  The analysis provided a comprehensive list of enriched terms across selected databases, highlighting significant GO. The results give a high-level summary of pathways or terms most relevant to the input data.

- **Detailed Results**:
  The detailed results section includes a tabulated format with enriched terms, adjusted p-values, and relevant statistics. Each entry provides information such as the enrichment score, associated genes, and functional annotations, allowing for an in-depth understanding of biological significance.

- **GO Context**:
  The Gene Ontology (GO) context is divided into three main categories: Biological Process (BP), Molecular Function (MF), and Cellular Component (CC). The analysis identifies which GO terms are significantly enriched, offering insights into the broader biological implications of the gene set. This helps in pinpointing processes such as cellular responses, metabolic pathways, and molecular interactions.

- **Query Info**:
  This section includes specifics about the input data, including the total number of queried genes and any identifiers not recognized or mapped. It also details the statistical background used, the chosen organism, and other analysis settings, ensuring transparency and reproducibility of the results.


####  {-}

#### Different Backgrounds

#### **Challenge:** How different backgrounds impact the output? {- .challenge}

Use 'All known genes' in one analysis and 'Custom' background in another. Download the results by clicking on CSV button. Browse the results in the spreadsheets and find out the difference between two.

#### **Question 1** {- .rationale}

Which background would you use in your analysis?

#### **Question 2** {- .rationale}

How can one perform Under Representation Analysis in gProfiler?

### Steps to perform GSEA in g:Profiler:
<span style="color:orange;">- Prepare Your Pre-ranked List:</span> Steps to provide a ranked gene list are given [here](degust.html).

<span style="color:orange;">- Input Gene List:</span> Paste your prepared gene list directly into the input box on the g:Profiler web page or upload a file containing your list.

<span style="color:orange;">- Select Organism:</span> Same as above.

<span style="color:orange;">- Select Ordered query:</span> The "Ordered query" option in g:Profiler is designed to work with pre-ranked gene lists.

<span style="color:orange;">- Set Significance Threshold:</span> Same as above.

<span style="color:orange;">- Provide a Custom GMT:</span> This GMT file can be downloaded from MSigDB.

<span style="color:orange;">- Run Query:</span> Same as above.

#### **Challenge:** GSEA with gProfiler {- .challenge}
Download the Hallmark gene sets (h.all.v2024.1.Hs.symbols.gmt) from [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp) and use it as Custom GMT.


## STRING
STRING (Search Tool for the Retrieval of Interacting Genes/Proteins) is a resource for exploring protein-protein interaction (PPI) networks. It combines experimental data, predictions, and curated information to build networks that highlight functional relationships, helping to reveal shared pathways or biological processes within gene or protein lists.

## Reactome
Reactome is an open-source database of curated biological pathways across species, offering pathway maps and enrichment tools to analyze gene lists in a pathway-focused context. Itâ€™s ideal for visualising data within established biochemical and cellular processes.

## MSigDB GSEA
MSigDB (Molecular Signatures Database) is a collection of gene sets for Gene Set Enrichment Analysis (GSEA), representing pathways and gene signatures linked to biological states or diseases. It helps identify enriched gene sets, aiding the analysis of gene expression changes and key pathways in experimental data.


