# Presenting reproducible FEA methods




Duration: 10 minutes 
Purpose of the activity: Promote awareness of and thinking about reproducibility in scientific methods, and what details from FEA are important for reproducibility. Encourage a space of active and collaborative participation to finish off the workshop, in order to promote participant confidence in asking questions. 


This will be some kind of real time document or PollEV where participants can type free text (with a word/character limit?) under the headings/sections of “Materials and Methods” and “Results”. All participants can see what others are typing but the responses are anonymous
During the webinar, we have raised the fact that reproducibility is key, because so many small changes to the analysis can change the results, and there are many peer reviewed articles that have critiqued a lack of reproducibility and serial under-reporting of methods in studies that report FEA. 
Ideal responses will be things like “include the name of the tool and its version”, “ multiple testing correction was performed with the Benjamni-Hochberg method and enrichments filted for adjusted P value < 0.05”, “the statistical background gene list is included in the supplementary methods” etc 
During/after real time participant responses, we can highlight common themes, add any that are missing from the generated list. 
I wonder if we could find a paper that has really awesome reproducible detail as an example and include the link in the material… 




https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009935


https://fairdomhub.org/investigations/583


https://ceur-ws.org/Vol-3415/paper-8.pdf

## GS version
# Reporting Results

The accuracy and clarity of methods and results reporting are crucial to the reproducibility of your findings. In functional enrichment analysis (FEA), each step has the potential to influence results (INSERT THAT PAPER THAT SAYS FEA IS OFTEN NOT REPORTED CORRECTLY). Therefore, a well-documented approach that includes detailed, version-controlled methods is essential for researchers to interpret, replicate, and build upon each other’s work.

This exercise will focus on crafting concise, structured reports for FEA methods and results. Using the example report provided below, along with our previous work in the (axolotl lesson)[link to axolotl lesson], identify areas where the reporting lacks clarity or detail. Pay close attention to the use of specific tools, parameter settings, and description of results to ensure that each part of the report is thorough and transparent.

## Mock methods writeup

We performed functional enrichment analysis the RNA-seq data from axolotl blastema samples to identify differentially expressed genes (DEGs) associated with limb regeneration. Gene Ontology (GO) annotations for the DEGs were retrieved using the online database BLAH, which provided access to the necessary GO terms. The enrichment analysis was performed with clusterProfiler to identify Biological Process (BP), Cellular Component (CC), and Molecular Function (MF) categories that were enriched in up- and down-regulated genes. We set the p-value threshold to 0.05 but also considered terms with slightly higher p-values. For visualisation, plots were generated to display the top enriched terms across GO categories in R.

### What is missing or incorrect? [POLL]

1. Lack of Specific Tools and Database names and versions
2. Missing parameters and cutoffs
3. Unclear visualisation methods

### Exemplar methods writeup 

INSERT METHODS WRITE UP

## Mock results writeup 

Our enrichment analysis revealed significant GO terms, particularly in the Biological Process category. These terms included processes related to tissue development and cellular growth. BLAH pathways were identified as enriched. Lastly, we compared enrichment results from the two databases to verify consistency. While some terms overlapped, there were notable differences in the results between the databases, including BLAH. 

### What is missing or incorrect? [POLL]

1. Specific GO terms, pathways and categories 
2. Details on the comparison of databases
3. Lack of quantitative information
4. Missing background context for consistency comparison

### Exemplar results writeup 

INSERT RESULTS WRITE UP

