<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Non-model species functional enrichment analysis</title>

<script src="novel_species_files/header-attrs-2.26/header-attrs.js"></script>
<script src="novel_species_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="novel_species_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="novel_species_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="novel_species_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="novel_species_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="novel_species_files/navigation-1.1/tabsets.js"></script>
<link href="novel_species_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="novel_species_files/highlightjs-9.12.0/highlight.js"></script>
<link href="novel_species_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="novel_species_files/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Non-model species functional enrichment
analysis</h1>

</div>


<div id="working-directory" class="section level1">
<h1>0. Working directory</h1>
<pre class="r"><code>getwd()</code></pre>
<pre><code>## [1] &quot;/home/user3/workshop&quot;</code></pre>
</div>
<div
id="import-emapper-axolotl-annotation-file-go-ontology-file-and-kegg-pathways-file"
class="section level1">
<h1>1. Import <code>emapper</code> axolotl annotation file, GO ontology
file and KEGG Pathways file</h1>
<div id="public-data-sources-used-for-this-notebook"
class="section level2">
<h2>1.1 Public data sources used for this notebook</h2>
<ul>
<li><a
href="https://www.axolotl-omics.org/dl/AmexG_v6.0-DD.fa.gz">Reference
genome</a></li>
<li><a
href="https://www.axolotl-omics.org/dl/AmexT_v47-AmexG_v6.0-DD.gtf.gz">GTF
file</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA300706">Raw
fastq</a></li>
<li><a href="https://purl.obolibrary.org/obo/go.obo">GO ‘core’ ontology
file</a></li>
</ul>
</div>
<div id="data-processing" class="section level2">
<h2>1.2 Data processing</h2>
<p>The raw reads were processed to counts matrix and DE results broadly
following <a href="https://github.com/Sydney-Informatics-Hub/RNASeq-DE"
class="uri">https://github.com/Sydney-Informatics-Hub/RNASeq-DE</a>
using the reference genome from www.axolotl-omics.org. Differential gene
expression analysis was performed in R with <code>DESeq2</code> v
1.46.0, filtering for genes with at least a count of 10 in at least 2
samples. The data comprises 2 groups (proximal blastema and distal
blastema) and 2 replicates per group.</p>
<p>A predicted proteome was created by extracting the predicted peptide
sequences from the GTF file from www.axolotl-omics.org then filtering
for longest isoform per gene with <code>AGAT</code> v 1.4.0. The
predicated proteome was annotated against GO and KEGG with
<code>eggNOG emapper</code> v 2.1.12.</p>
</div>
<div id="import-annotation-files" class="section level2">
<h2>1.3 Import annotation files</h2>
<div id="emapper-proteome-annotation" class="section level3">
<h3>1.3.1 emapper proteome annotation</h3>
<p>The <code>emapper</code> annotation output contains results against a
number of databases including GO and KEGG, which we will focus on
today.</p>
<pre class="r"><code>eggnog_anno &lt;- read_tsv(&quot;AmexG_v6.0-DD.emapper.annotations.txt&quot;, show_col_types = FALSE) 
head(eggnog_anno)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["#query"],"name":[1],"type":["chr"],"align":["left"]},{"label":["seed_ortholog"],"name":[2],"type":["chr"],"align":["left"]},{"label":["evalue"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["score"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["eggNOG_OGs"],"name":[5],"type":["chr"],"align":["left"]},{"label":["max_annot_lvl"],"name":[6],"type":["chr"],"align":["left"]},{"label":["COG_category"],"name":[7],"type":["chr"],"align":["left"]},{"label":["Description"],"name":[8],"type":["chr"],"align":["left"]},{"label":["Preferred_name"],"name":[9],"type":["chr"],"align":["left"]},{"label":["GOs"],"name":[10],"type":["chr"],"align":["left"]},{"label":["EC"],"name":[11],"type":["chr"],"align":["left"]},{"label":["KEGG_ko"],"name":[12],"type":["chr"],"align":["left"]},{"label":["KEGG_Pathway"],"name":[13],"type":["chr"],"align":["left"]},{"label":["KEGG_Module"],"name":[14],"type":["chr"],"align":["left"]},{"label":["KEGG_Reaction"],"name":[15],"type":["chr"],"align":["left"]},{"label":["KEGG_rclass"],"name":[16],"type":["chr"],"align":["left"]},{"label":["BRITE"],"name":[17],"type":["chr"],"align":["left"]},{"label":["KEGG_TC"],"name":[18],"type":["chr"],"align":["left"]},{"label":["CAZy"],"name":[19],"type":["chr"],"align":["left"]},{"label":["BiGG_Reaction"],"name":[20],"type":["chr"],"align":["left"]},{"label":["PFAMs"],"name":[21],"type":["chr"],"align":["left"]}],"data":[{"1":"AMEX60DD012313","2":"8479.XP_005296626.1","3":"2.93e-32","4":"122.0","5":"COG2801@1|root,KOG0017@2759|Eukaryota,38F42@33154|Opisthokonta,3BA5H@33208|Metazoa,3CWAK@33213|Bilateria,48B0V@7711|Chordata,499U9@7742|Vertebrata","6":"33208|Metazoa","7":"E","8":"K02A2.6-like","9":"-","10":"-","11":"-","12":"-","13":"-","14":"-","15":"-","16":"-","17":"-","18":"-","19":"-","20":"-","21":"RVT_1,SCAN,rve,zf-CCHC,zf-H2C2"},{"1":"AMEX60DD012322","2":"7668.SPU_016857-tr","3":"3.52e-24","4":"104.0","5":"KOG1075@1|root,KOG1075@2759|Eukaryota,39N4Q@33154|Opisthokonta,3B9WN@33208|Metazoa,3D1S8@33213|Bilateria","6":"33208|Metazoa","7":"L","8":"Reverse transcriptase (RNA-dependent DNA polymerase)","9":"-","10":"-","11":"-","12":"-","13":"-","14":"-","15":"-","16":"-","17":"-","18":"-","19":"-","20":"-","21":"RVT_1"},{"1":"AMEX60DD012333","2":"42254.XP_004616260.1","3":"2.07e-56","4":"177.0","5":"2CZ0H@1|root,2S7N2@2759|Eukaryota,3A9CP@33154|Opisthokonta,3BUHK@33208|Metazoa,3DBP1@33213|Bilateria,48GRW@7711|Chordata,49DK1@7742|Vertebrata,3JI75@40674|Mammalia","6":"33208|Metazoa","7":"-","8":"-","9":"-","10":"-","11":"-","12":"-","13":"-","14":"-","15":"-","16":"-","17":"-","18":"-","19":"-","20":"-","21":"-"},{"1":"AMEX60DD012336","2":"38654.XP_006038865.1","3":"2.90e-06","4":"49.7","5":"KOG1721@1|root,KOG1721@2759|Eukaryota,38BZ9@33154|Opisthokonta,3BA9A@33208|Metazoa,3CT35@33213|Bilateria,4824W@7711|Chordata","6":"33208|Metazoa","7":"C","8":"nucleic acid binding","9":"-","10":"-","11":"-","12":"ko:K09228,ko:K09230","13":"-","14":"-","15":"-","16":"-","17":"ko00000,ko03000","18":"-","19":"-","20":"-","21":"KRAB,zf-C2H2,zf-C2H2_6"},{"1":"AMEX60DD012351","2":"13735.ENSPSIP00000002556","3":"2.49e-08","4":"64.7","5":"KOG1721@1|root,KOG1721@2759|Eukaryota,38KXR@33154|Opisthokonta,3BFID@33208|Metazoa,3D395@33213|Bilateria,489GW@7711|Chordata,4965T@7742|Vertebrata,4CH9P@8459|Testudines","6":"33208|Metazoa","7":"S","8":"Zinc finger protein","9":"-","10":"-","11":"-","12":"ko:K09228","13":"-","14":"-","15":"-","16":"-","17":"ko00000,ko03000","18":"-","19":"-","20":"-","21":"DUF3669,KRAB,zf-C2H2"},{"1":"AMEX60DD012352","2":"69293.ENSGACP00000013829","3":"6.57e-19","4":"85.1","5":"2CW8H@1|root,2RTVJ@2759|Eukaryota,3ANVN@33154|Opisthokonta,3C1UE@33208|Metazoa,3DHUM@33213|Bilateria,48K6V@7711|Chordata,49H6N@7742|Vertebrata,4A7WQ@7898|Actinopterygii","6":"33208|Metazoa","7":"-","8":"-","9":"-","10":"-","11":"-","12":"-","13":"-","14":"-","15":"-","16":"-","17":"-","18":"-","19":"-","20":"-","21":"-"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>The raw annotation file provides us with ‘term ID to gene ID’
mappings for our species. We also need ‘term ID to term description’
mappings. These files are not organisms specific: we will extract only
the terms that are found within our custom species annotation, to make
our organism specific version.</p>
</div>
<div id="go-core-ontology" class="section level3">
<h3>1.3.2 GO Core Ontology</h3>
<p>For GO, we will use the GO ‘core’ ontology file, downloaded from <a
href="https://purl.obolibrary.org/obo/go.obo"
class="uri">https://purl.obolibrary.org/obo/go.obo</a> and included in
the data files you downloaded to <code>workshop</code> directory
earlier.</p>
<p>We will use the <code>ontologyIndex</code> package to retrieve
ontologi info and save to an object named <code>ontology</code> for
later use creating the required custom database files for
<code>clusterProfiler</code> and <code>WebGestaltR</code>.</p>
<pre class="r"><code>ontology &lt;- ontologyIndex::get_ontology(file = &quot;go.obo&quot;,
  propagate_relationships = &quot;is_a&quot;, #propagates relationships from parent terms to children
  extract_tags = &quot;everything&quot;, # retrieve all available details for each term
  merge_equivalent_terms = TRUE) # avoid unecessary redundancy </code></pre>
</div>
<div id="kegg-pathways" class="section level3">
<h3>1.3.3 KEGG Pathways</h3>
<p>For KEGG, we have both <code>map</code> and <code>ko</code> IDs in
our <code>emapper</code> annotation.</p>
<p><code>ko</code> terms (<a href="https://www.genome.jp/kegg/ko.html"
class="uri">https://www.genome.jp/kegg/ko.html</a>) represent
orthologous groups of genes, which are assigned based on evolutionary
relationships and functional similarity, so can provide more precise
functional categorisation which can be particularly useful when working
with novel species which lack curate dpathway information.</p>
<p><code>map</code> terms (<a
href="https://www.genome.jp/kegg/pathway.html"
class="uri">https://www.genome.jp/kegg/pathway.html</a>) are manually
drawn pathway maps representing KEGG database of molecular interaction,
reaction and relation networks for: Metabolism, Genetic Information
Processing, Environmental Information Processing, Cellular Processes,
Organismal Systems, Human Diseases, and Drug Development.</p>
<p>Today we will be working with the <code>map</code> terms due to
database download restrictions.</p>
<p>Free access to the KEGG FTP downloads requires an academic
subscription, to which you must confirm to be the “only user of the KEGG
FTP Data”. The pathways list was available freely. As a single user, you
can request academic access here <a
href="https://www.pathway.jp/en/academic.html"
class="uri">https://www.pathway.jp/en/academic.html</a>.</p>
<p>There is an alternate method for using <code>ko</code> IDs, which
uses the KEGG ontology information available through the
<code>clusterProfiler</code> functions <code>enrichKEGG</code> and
<code>gseKEGG</code>. An example of the R code can be found here <a
href="https://github.com/dadrasarmin/enrichment_analysis_for_non_model_organism"
class="uri">https://github.com/dadrasarmin/enrichment_analysis_for_non_model_organism</a>.
However, this poses a problem: as the novel species gene IDs are
assigned to KEGG terms, gene:name duplicate records are identified, and
duplicates must be removed in order to avoid errors running the
enrichment. This loss of data will have a real impact on the results,
with the importance of some terms being underestimated.</p>
<p>Given these considerations, we will proceed with <code>map</code>
pathway terms :-)</p>
<p>The KEGG <code>map</code> pathway list was downloaded from <a
href="https://rest.kegg.jp/list/pathway"
class="uri">https://rest.kegg.jp/list/pathway</a> and downloaded to your
<code>workshop</code> directory.</p>
<pre class="r"><code>kegg_pathways &lt;- read.table(&quot;kegg_pathways_2024-11-13.txt&quot;, header = FALSE, sep = &quot;\t&quot;, col.names = c(&quot;term&quot;, &quot;name&quot;))
head(kegg_pathways)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"map01100","2":"Metabolic pathways","_rn_":"1"},{"1":"map01110","2":"Biosynthesis of secondary metabolites","_rn_":"2"},{"1":"map01120","2":"Microbial metabolism in diverse environments","_rn_":"3"},{"1":"map01200","2":"Carbon metabolism","_rn_":"4"},{"1":"map01210","2":"2-Oxocarboxylic acid metabolism","_rn_":"5"},{"1":"map01212","2":"Fatty acid metabolism","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</div>
<div
id="import-axolotl-de-results-file-and-extract-gene-lists-for-ora-and-gsea"
class="section level1">
<h1>2. Import axolotl DE results file and extract gene lists for ORA and
GSEA</h1>
<div id="import-axolotl-de-data" class="section level2">
<h2>2.1 Import axolotl DE data</h2>
<p>Load the DE results file for axolotl:</p>
<pre class="r"><code>de_matrix &lt;- read_tsv(&quot;axolotl_DE_results.txt&quot;, col_names = TRUE, show_col_types = FALSE)
head(de_matrix)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["geneID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["baseMean"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["log2FoldChange"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["lfcSE"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["stat"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["pvalue"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["padj"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"AMEX60DD000001","2":"38.75552","3":"0.16001768","4":"0.7487504","5":"0.21371299","6":"0.8307709","7":"0.9999735"},{"1":"AMEX60DD000002","2":"74.50889","3":"-0.18676860","4":"0.5309943","5":"-0.35173368","6":"0.7250380","7":"0.9999735"},{"1":"AMEX60DD000003","2":"519.72598","3":"-0.16751319","4":"0.3189346","5":"-0.52522747","6":"0.5994251","7":"0.9999735"},{"1":"AMEX60DD000004","2":"235.65733","3":"-0.16228436","4":"0.3764389","5":"-0.43110410","6":"0.6663927","7":"0.9999735"},{"1":"AMEX60DD000005","2":"130.15166","3":"-0.13738690","4":"0.4190395","5":"-0.32786147","6":"0.7430164","7":"0.9999735"},{"1":"AMEX60DD000006","2":"68.15811","3":"-0.05596896","4":"0.6071422","5":"-0.09218427","6":"0.9265516","7":"0.9999735"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="create-the-ranked-gene-list-for-gsea" class="section level2">
<h2>2.2 Create the ranked gene list for GSEA</h2>
<p>Recall from the last 2 activities that <code>clusterProfiler</code>
requires a vector object for GSEA, while <code>WebGestaltR</code>
requires a 2-column dataframe. Since we intend to use both tools, let’s
create both now:</p>
<pre class="r"><code># Create ranked vector for clusterProfiler GSEA
ranked_vector &lt;- setNames(de_matrix$log2FoldChange, de_matrix$geneID) %&gt;% sort(decreasing = TRUE)  # Named vector

# check
head(ranked_vector)</code></pre>
<pre><code>## AMEX60DD020778 AMEX60DD020772 AMEX60DD020773 AMEX60DD005693 AMEX60DD020780 
##      10.596655      10.583791      10.281349      10.067803       9.789798 
## AMEX60DD030633 
##       9.691689</code></pre>
<pre class="r"><code>tail(ranked_vector)</code></pre>
<pre><code>## AMEX60DD051201 AMEX60DD028124 AMEX60DD044496 AMEX60DD020182 AMEX60DD053589 
##      -7.886591      -7.939796      -8.070068      -8.255430      -9.226586 
## AMEX60DD007432 
##      -9.918767</code></pre>
<pre class="r"><code># Create ranked dataframe for WebGestaltR GSEA# extract ranked dataframe
ranked_df &lt;- de_matrix %&gt;%
  arrange(desc(log2FoldChange)) %&gt;%
  dplyr::select(geneID, log2FoldChange)

# check
head(ranked_df)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["geneID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["log2FoldChange"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"AMEX60DD020778","2":"10.596655"},{"1":"AMEX60DD020772","2":"10.583791"},{"1":"AMEX60DD020773","2":"10.281349"},{"1":"AMEX60DD005693","2":"10.067803"},{"1":"AMEX60DD020780","2":"9.789798"},{"1":"AMEX60DD030633","2":"9.691689"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>tail(ranked_df)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["geneID"],"name":[1],"type":["chr"],"align":["left"]},{"label":["log2FoldChange"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"AMEX60DD051201","2":"-7.886591"},{"1":"AMEX60DD028124","2":"-7.939796"},{"1":"AMEX60DD044496","2":"-8.070068"},{"1":"AMEX60DD020182","2":"-8.255430"},{"1":"AMEX60DD053589","2":"-9.226586"},{"1":"AMEX60DD007432","2":"-9.918767"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="create-gene-lists-for-ora" class="section level2">
<h2>2.3 Create gene lists for ORA</h2>
<p>For ORA, both tools require vector class gene lists. We will filter
for adjusted P value &lt; 0.01 and absolute log2 fold change greater
than 1.5. The matrix has already filtered out genes with very low counts
so we take all genes present as the background.</p>
<pre class="r"><code># Filter for DEGs and save gene IDs as vector  
degs &lt;- de_matrix %&gt;%
  filter(padj &lt;= 0.01 &amp; abs(log2FoldChange) &gt;= 1.5) %&gt;%
  pull(geneID)  # Extract 

# Extract the background gene list vector 
background &lt;- de_matrix %&gt;%
  pull(geneID)  

# Check number of genes: 
cat(&quot;Number of DEGs:&quot;, length(degs), &quot;\n&quot;)         # Number of DEGs</code></pre>
<pre><code>## Number of DEGs: 247</code></pre>
<pre class="r"><code>cat(&quot;Number of background genes:&quot;, length(background), &quot;\n&quot;)   # Number of background genes</code></pre>
<pre><code>## Number of background genes: 24419</code></pre>
<pre class="r"><code># Check format: 
head(degs)</code></pre>
<pre><code>## [1] &quot;AMEX60DD000080&quot; &quot;AMEX60DD000147&quot; &quot;AMEX60DD001144&quot; &quot;AMEX60DD001307&quot;
## [5] &quot;AMEX60DD001377&quot; &quot;AMEX60DD001828&quot;</code></pre>
<pre class="r"><code>head(background)</code></pre>
<pre><code>## [1] &quot;AMEX60DD000001&quot; &quot;AMEX60DD000002&quot; &quot;AMEX60DD000003&quot; &quot;AMEX60DD000004&quot;
## [5] &quot;AMEX60DD000005&quot; &quot;AMEX60DD000006&quot;</code></pre>
</div>
<div id="save-gene-lists" class="section level2">
<h2>2.4 Save gene lists</h2>
<p>Saving any outputs generated from R code is vital to reproducibility!
You should include all analysed gene lists within the supplementary
materials of your manuscript.</p>
<pre class="r"><code># Save DEGs
write.table(degs, file = &quot;Axolotl_DEGs.txt&quot;, quote = FALSE, col.names = FALSE, row.names = FALSE, sep = &quot;\t&quot;)
# Save background
write.table(background, file = &quot;Axolotl_background.txt&quot;, quote = FALSE, col.names = FALSE, row.names = FALSE, sep = &quot;\t&quot;)
# Save ranked
write.table(ranked_df, file = &quot;Axolotl_rankedFC.txt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = TRUE)</code></pre>
</div>
</div>
<div
id="reformat-annotation-files-for-clusterprofiler-go-and-kegg-analysis"
class="section level1">
<h1>3. Reformat annotation files for <code>clusterProfiler</code> GO and
KEGG analysis</h1>
<div id="create-term2gene-files" class="section level2">
<h2>3.1 Create TERM2GENE files</h2>
<p>These are 2 column text files with the term ID (one per line)
alongside the ID of the gene that maps to the term. A gene can map to
many terms and thus be present on multiple lines. A term can be mapped
to more than one gene and thus be present on many lines.</p>
<p>Check the column names of the <code>emapper</code> annotation file so
we know which are the GO and KEGG column names:</p>
<pre class="r"><code>colnames(eggnog_anno)</code></pre>
<pre><code>##  [1] &quot;#query&quot;         &quot;seed_ortholog&quot;  &quot;evalue&quot;         &quot;score&quot;         
##  [5] &quot;eggNOG_OGs&quot;     &quot;max_annot_lvl&quot;  &quot;COG_category&quot;   &quot;Description&quot;   
##  [9] &quot;Preferred_name&quot; &quot;GOs&quot;            &quot;EC&quot;             &quot;KEGG_ko&quot;       
## [13] &quot;KEGG_Pathway&quot;   &quot;KEGG_Module&quot;    &quot;KEGG_Reaction&quot;  &quot;KEGG_rclass&quot;   
## [17] &quot;BRITE&quot;          &quot;KEGG_TC&quot;        &quot;CAZy&quot;           &quot;BiGG_Reaction&quot; 
## [21] &quot;PFAMs&quot;</code></pre>
<p>We need <code>GOs</code> and <code>KEGG_Pathway</code> columns.</p>
<div id="go-term2gene" class="section level3">
<h3>3.1.1 GO TERM2GENE</h3>
<pre class="r"><code># Extract GO IDs from eggNOG annotation file to correct format for clusterProfiler

go_term2gene &lt;- eggnog_anno %&gt;%
    dplyr::select(GOs, `#query`) %&gt;% # select the GO column and the query column (axolotl gene ID) 
    dplyr::filter(GOs != &quot;-&quot;) %&gt;% # filter out rows where the GO ID is &quot;-&quot; ie no GO annotation for this gene
    separate_rows(GOs, sep = &quot;,&quot;) %&gt;% # split comma-delimited list of many GO terms for a gene into separate rows
    dplyr::select(GOs, `#query`) %&gt;% # keep the GO and query columns
    distinct() %&gt;% # remove any duplicate rows 
    drop_na() # remove rows with missing values

# Rename columns to match desired output format
colnames(go_term2gene) &lt;- c(&quot;term&quot;, &quot;gene&quot;)

# Save to file
write.table(go_term2gene, file = &quot;Axolotl_GO_term2gene.txt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = TRUE)

# Number of rows: 
cat(&quot;Number of GO term2gene rows:&quot;, nrow(go_term2gene), &quot;\n&quot;)</code></pre>
<pre><code>## Number of GO term2gene rows: 2839193</code></pre>
<pre class="r"><code># Check first few rows
head(go_term2gene)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["gene"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"GO:0000981","2":"AMEX60DD012362"},{"1":"GO:0003674","2":"AMEX60DD012362"},{"1":"GO:0003700","2":"AMEX60DD012362"},{"1":"GO:0006355","2":"AMEX60DD012362"},{"1":"GO:0006357","2":"AMEX60DD012362"},{"1":"GO:0008150","2":"AMEX60DD012362"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="kegg-term2gene" class="section level3">
<h3>3.1.2 KEGG TERM2GENE</h3>
<p>Here we use the same process as above, just with a different column
name.</p>
<pre class="r"><code>kegg_term2gene &lt;- eggnog_anno %&gt;%
    dplyr::select(KEGG_Pathway, `#query`) %&gt;%  # Select the relevant columns
    dplyr::filter(grepl(&quot;map&quot;, KEGG_Pathway)) %&gt;%  # Keep only rows where KEGG_Pathway contains &#39;map&#39;
    separate_rows(KEGG_Pathway, sep = &quot;,&quot;) %&gt;%  # Split multiple pathways into separate rows
    dplyr::mutate(term = gsub(&quot;map:&quot;, &quot;&quot;, KEGG_Pathway)) %&gt;%  # Remove the &quot;map:&quot; prefix
    dplyr::filter(grepl(&quot;^map&quot;, term)) %&gt;%  # Filter again to make sure we only have map pathways (after removing &quot;map:&quot;)
    dplyr::select(term, `#query`) %&gt;%  # Select the pathway (term) and gene columns
    distinct() %&gt;%  # Remove duplicate rows
    drop_na()  # Remove rows with missing values


# Rename columns to match desired output format
colnames(kegg_term2gene) &lt;- c(&quot;term&quot;, &quot;gene&quot;)

# Save to file
write.table(kegg_term2gene, file = &quot;Axolotl_KEGG-Pathways_term2gene.txt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = TRUE)

cat(&quot;Number of KEGG term2gene rows:&quot;, nrow(kegg_term2gene), &quot;\n&quot;)</code></pre>
<pre><code>## Number of KEGG term2gene rows: 59305</code></pre>
<pre class="r"><code># View result to check
head(kegg_term2gene)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["gene"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"map04621","2":"AMEX60DD012519"},{"1":"map04740","2":"AMEX60DD012537"},{"1":"map04137","2":"AMEX60DD012545"},{"1":"map04139","2":"AMEX60DD012545"},{"1":"map04514","2":"AMEX60DD012547"},{"1":"map04530","2":"AMEX60DD012547"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div id="term2name" class="section level2">
<h2>3.2 TERM2NAME</h2>
<div id="go" class="section level3">
<h3>3.2.1 GO</h3>
<p>The GO Core Ontology file (we have saved in our <code>workshop</code>
folder and labeleld the filename as R object <code>go_core</code>) will
enable us to assign term descriptions to term IDs and create our
<code>TERM2NAME</code> files.</p>
<p>This may take a few moments to run. It will use the
<code>ontology</code> object we created earlier.</p>
<pre class="r"><code># Create term to name table, removing duplicates, missing values and obsolete terms 
go_term2name &lt;- go_term2gene %&gt;%
    mutate(name = ontology$name[term]) %&gt;%
    dplyr::select(term, name) %&gt;%
    distinct() %&gt;%
    drop_na() %&gt;%
    filter(!grepl(&quot;obsolete&quot;, name))

# Save to file
write.table(go_term2name, file = &quot;Axolotl_GO_term2name.txt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = TRUE)

# Show the first few lines
head(go_term2name)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"GO:0000981","2":"DNA-binding transcription factor activity, RNA polymerase II-specific"},{"1":"GO:0003674","2":"molecular_function"},{"1":"GO:0003700","2":"DNA-binding transcription factor activity"},{"1":"GO:0006355","2":"regulation of DNA-templated transcription"},{"1":"GO:0006357","2":"regulation of transcription by RNA polymerase II"},{"1":"GO:0008150","2":"biological_process"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="kegg" class="section level3">
<h3>3.2.2 KEGG</h3>
<p>The KEGG Pathways file was available for download in the correct
format for <code>TERM2NAME</code>.</p>
<pre class="r"><code>head(kegg_pathways)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"map01100","2":"Metabolic pathways","_rn_":"1"},{"1":"map01110","2":"Biosynthesis of secondary metabolites","_rn_":"2"},{"1":"map01120","2":"Microbial metabolism in diverse environments","_rn_":"3"},{"1":"map01200","2":"Carbon metabolism","_rn_":"4"},{"1":"map01210","2":"2-Oxocarboxylic acid metabolism","_rn_":"5"},{"1":"map01212","2":"Fatty acid metabolism","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Let’s restrict it to include the terms relevant to our analysis, and
then print that to a file for reproducibility.</p>
<pre class="r"><code>kegg_term2name &lt;- kegg_pathways %&gt;%
  dplyr::filter(term %in% kegg_term2gene$term) %&gt;%  # Only keep terms that are in kegg_term2gene
  distinct() %&gt;%  # Remove duplicate entries
  drop_na()  # Remove rows with missing values

# Save the result to a file
write.table(kegg_term2name, file = &quot;Axolotl_KEGG_term2name.txt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = TRUE)

# Check first few rows
head(kegg_term2name)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"map01100","2":"Metabolic pathways","_rn_":"1"},{"1":"map01110","2":"Biosynthesis of secondary metabolites","_rn_":"2"},{"1":"map01120","2":"Microbial metabolism in diverse environments","_rn_":"3"},{"1":"map01200","2":"Carbon metabolism","_rn_":"4"},{"1":"map01210","2":"2-Oxocarboxylic acid metabolism","_rn_":"5"},{"1":"map01212","2":"Fatty acid metabolism","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div id="count-annotations" class="section level2">
<h2>3.3 Count annotations</h2>
<p>How much of our proteome was annotated? What about our DEGs and
background? Genes that do not have any annotation are excluded from
enrichment analysis, so having an understanding of the extent of
annotation for your novel species is very important when interpreting
results!</p>
<p>Count the number of GO terms found within the genome, and the number
of genes with GO annotations:</p>
<pre class="r"><code>go_total_terms&lt;-nrow(go_term2gene)
print(paste(&quot;Total annotations to GO:&quot;, go_total_terms))</code></pre>
<pre><code>## [1] &quot;Total annotations to GO: 2839193&quot;</code></pre>
<pre class="r"><code>go_unique_genes &lt;- length(unique(go_term2gene$gene))
print(paste(&quot;Number of unique genes with 1 or more annotation terms:&quot;, go_unique_genes))</code></pre>
<pre><code>## [1] &quot;Number of unique genes with 1 or more annotation terms: 21373&quot;</code></pre>
<p>And for KEGG:</p>
<pre class="r"><code>kegg_total_terms&lt;-nrow(kegg_term2gene)
print(paste(&quot;Total annotations to KEGG Pathways:&quot;, kegg_total_terms))</code></pre>
<pre><code>## [1] &quot;Total annotations to KEGG Pathways: 59305&quot;</code></pre>
<pre class="r"><code>kegg_unique_genes &lt;- length(unique(kegg_term2gene$gene))
print(paste(&quot;Number of unique genes with 1 or more annotation terms:&quot;, kegg_unique_genes))</code></pre>
<pre><code>## [1] &quot;Number of unique genes with 1 or more annotation terms: 12226&quot;</code></pre>
<p>47,196 putative axolotl proteins were annotated. That’s around 1/4 of
our genes mapped to KEGG Pathways, and less than half of our genes
mapped to GO! Ouch. As much as we expect this with uncurated novel
species genomes, it’s still unpleasant to face :-)</p>
<p>What of the genes in our gene list specifically? We have an uncurated
proteome, yet the genes in our input matrix were expressed at a
meaningful level within axolotl, so these may actually have a higher
annotation percentage than all genes in the proteome.</p>
<pre class="r"><code># Filter the term2gene table to only include genes in the background gene list
go_filtered_term2gene &lt;- go_term2gene %&gt;% filter(gene %in% background)

# Count the number of unique background genes with at least one GO term
unique_genes_with_go &lt;- go_filtered_term2gene %&gt;% distinct(gene) %&gt;% nrow()

# Calculate the percentage of background genes that have GO annotations
percent_go_unique &lt;- (unique_genes_with_go / length(background)) * 100

# Print results
cat(&quot;Number of input genes with GO annotations:&quot;, unique_genes_with_go, &quot;(&quot;,percent_go_unique,&quot;%)\n&quot;)</code></pre>
<pre><code>## Number of input genes with GO annotations: 15072 ( 61.72243 %)</code></pre>
<pre class="r"><code># Filter the term2gene table to only include genes in the background gene list
kegg_filtered_term2gene &lt;- kegg_term2gene %&gt;% filter(gene %in% background)

# Count the number of unique background genes with at least one GO term
unique_genes_with_kegg &lt;- kegg_filtered_term2gene %&gt;% distinct(gene) %&gt;% nrow()

# Calculate the percentage of background genes that have GO annotations
percent_kegg_unique &lt;- (unique_genes_with_kegg / length(background)) * 100

# Print results
cat(&quot;Number of input genes with KEGG Pathways annotations:&quot;, unique_genes_with_kegg, &quot;(&quot;,percent_kegg_unique,&quot;%)\n&quot;)</code></pre>
<pre><code>## Number of input genes with KEGG Pathways annotations: 8000 ( 32.76137 %)</code></pre>
<p>As expected, the annotation % is higher for expressed genes than all
predicted genes, and very much higher than the GTF of 99,088 predicted
gene models (!!!) with an annotation rate of 21.6%.</p>
<p>This highlights a major caveat when performing FEA on non-model
species: the results are only as good as the annotations behind them.
Therefore, all results must be interpreted with caution. For many novel
(and under-funded) species, there are little opportunities (at present)
to improve the annotation. Some in-silico predicted genes appear to be
highly expressed and significantly regulated yet have no significant
similarity to anything in the non-redundant nucleotide or protein
databases. When working with datasets like this, it is critical to
explore those individual genes through other methods, in addition to
trying to garner some higher level overview such as we aim to obtain
from FEA. Hopefully, recent advances in AI protein modelling can help
provide insights into the functions of these novel genes.</p>
<p>For the axolotl with only 22% of predicted genes annotated, its clear
that the in-silico gene predictions within the GTF file require much
curation!</p>
</div>
</div>
<div id="run-clusterprofiler-universal-fea-functions-enricher-and-gsea"
class="section level1">
<h1>4. Run <code>clusterProfiler</code> universal FEA functions
<code>enricher</code> and <code>GSEA</code></h1>
<p>In the interest of time, and to try and cover as many options as
possible, let’s do ORA with GO and GSEA with KEGG for both tools.</p>
<div id="clusterprofiler-ora-of-go-terms" class="section level2">
<h2>4.1 clusterProfiler ORA of GO terms</h2>
<p>The <code>enricher</code> function is the ‘universal’ ORA option that
accepts the <code>TERM2GENE</code> and <code>TERM2NAME</code> files we
have just created.</p>
<p>Let’s review the help page:</p>
<pre class="r"><code>?clusterProfiler::enricher</code></pre>
<p>There are parameters for both adjusted P value and q value. Terms
must pass all thresholds (unadjusted P, adjusted P, and q value) so the
important filter will be the most stringent test applied. Let’s go with
BH and 0.05 which we have used regularly within this workshop and are
fairly common choices in the field.</p>
<pre class="r"><code>cp_go_ora &lt;- enricher(
  gene = degs,
  pvalueCutoff = 0.05,
  pAdjustMethod = &quot;BH&quot;,
  universe = background,
  minGSSize = 10,
  maxGSSize = 500,
  TERM2GENE = go_term2gene,
  TERM2NAME = go_term2name
)</code></pre>
<pre class="r"><code>cp_go_ora</code></pre>
<pre><code>## #
## # over-representation test
## #
## #...@organism     UNKNOWN 
## #...@ontology     UNKNOWN 
## #...@gene     chr [1:247] &quot;AMEX60DD000080&quot; &quot;AMEX60DD000147&quot; &quot;AMEX60DD001144&quot; ...
## #...pvalues adjusted by &#39;BH&#39; with cutoff &lt;0.05 
## #...91 enriched terms found
## &#39;data.frame&#39;:    91 obs. of  12 variables:
##  $ ID            : chr  &quot;GO:0048821&quot; &quot;GO:0070268&quot; &quot;GO:0031424&quot; &quot;GO:0019317&quot; ...
##  $ Description   : chr  &quot;erythrocyte development&quot; &quot;cornification&quot; &quot;keratinization&quot; &quot;fucose catabolic process&quot; ...
##  $ GeneRatio     : chr  &quot;7/145&quot; &quot;7/145&quot; &quot;7/145&quot; &quot;5/145&quot; ...
##  $ BgRatio       : chr  &quot;45/15072&quot; &quot;46/15072&quot; &quot;48/15072&quot; &quot;18/15072&quot; ...
##  $ RichFactor    : num  0.156 0.152 0.146 0.278 0.278 ...
##  $ FoldEnrichment: num  16.2 15.8 15.2 28.9 28.9 ...
##  $ zScore        : num  10.04 9.92 9.68 11.66 11.66 ...
##  $ pvalue        : num  2.20e-07 2.58e-07 3.49e-07 5.96e-07 5.96e-07 ...
##  $ p.adjust      : num  0.000266 0.000266 0.000266 0.000266 0.000266 ...
##  $ qvalue        : num  0.000242 0.000242 0.000242 0.000242 0.000242 ...
##  $ geneID        : chr  &quot;AMEX60DD002868/AMEX60DD025537/AMEX60DD026264/AMEX60DD026267/AMEX60DD032898/AMEX60DD032958/AMEX60DD032960&quot; &quot;AMEX60DD004554/AMEX60DD010118/AMEX60DD016775/AMEX60DD038318/AMEX60DD039599/AMEX60DD039603/AMEX60DD039606&quot; &quot;AMEX60DD004554/AMEX60DD010118/AMEX60DD016775/AMEX60DD038318/AMEX60DD039599/AMEX60DD039603/AMEX60DD039606&quot; &quot;AMEX60DD004640/AMEX60DD004644/AMEX60DD004647/AMEX60DD004651/AMEX60DD017241&quot; ...
##  $ Count         : int  7 7 7 5 5 5 16 5 5 7 ...
## #...Citation
## S Xu, E Hu, Y Cai, Z Xie, X Luo, L Zhan, W Tang, Q Wang, B Liu, R Wang, W Xie, T Wu, L Xie, G Yu. Using clusterProfiler to characterize multiomics data. Nature Protocols. 2024, doi:10.1038/s41596-024-01020-z</code></pre>
<p>91 significantly enriched terms at P.adj &lt; 0.05.</p>
<p>Note that we used the gene list object <code>degs</code> which had
247 genes, but the tool has applied the input size as 145 - this is
because it is automatically discarding any that do not have annotations.
Results would be the same if we instead used <code>annotated_degs</code>
object. Likewise, the background size is being reported as 15072 (the
number annotated) not 24,419 (the total in background list).</p>
<p>Save the results to a text file:</p>
<pre class="r"><code>file &lt;- &quot;Axolotl_clusterProfiler_GO_ORA_results.tsv&quot;
write.table(cp_go_ora, file, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE)  </code></pre>
<p>Let’s visualise with one of my favourite <code>enrichplot</code>
plots, the treeplot! Another advantage of this plot is that it can be
used for both ORA and GSEA results, so we can compare more easily. We
will add a custom subtitle that informs the number of DEGs that were
actually annotated and included in the FEA, so anyone reviewing the plot
will understand that caution must be exercised when interpreting the
results.</p>
<pre class="r"><code># calculate pairwise similarities between the enriched terms
cp_go_ora &lt;- enrichplot::pairwise_termsim(cp_go_ora)
p&lt;- enrichplot::treeplot(cp_go_ora, 
  showCategory = 15, 
  color = &quot;p.adjust&quot;, 
  cluster.params = list(label_words_n = 5)
)

# Add annotations (number of input genes and number of input genes with GO terms)
num_genes &lt;- length(degs)
genes_with_GO_terms &lt;- sum(degs %in% go_term2gene$gene)

# Print the plot with custom sub-title
p &lt;- p + ggtitle(&quot;clusterProfiler ORA of GO terms&quot;) + labs(subtitle = paste(&quot;Input genes:&quot;, num_genes, &quot;| Input genes with GO terms:&quot;, genes_with_GO_terms))
print(p)</code></pre>
<p><img src="novel_species_files/figure-html/treeplot%20CP%20GO%20ORA,%20-1.png" width="1152" /></p>
<p>There’s a lot of skin and muscle stuff, which we expect to be
expressed in the blastema. As for why they are dysregulated? This is a
dummy experiment from public RNAseq, with poor replication, and may not
even be the right experiment type for this question, so let’s not hope
for too many clear answers :-)</p>
</div>
<div id="clusterprofiler-gsea-of-kegg-terms" class="section level2">
<h2>4.2 clusterProfiler GSEA of KEGG terms</h2>
<p>The <code>GSEA</code> function is the ‘universal’ GSEA option that
accepts the <code>TERM2GENE</code> and <code>TERM2NAME</code> files we
have just created.</p>
<p>Let’s review the help page:</p>
<pre class="r"><code>?clusterProfiler::GSEA</code></pre>
<p>Recall from our <code>clusterProfiler</code> session with human data
that we needed to add <code>nPermSimple = 10000</code> to avoid a
warning about “unbalanced (positive and negative) gene-level statistic
value” and reduce <code>eps</code> to zero to avoid a warning about
obtaining better P value estiamtes . Let’s do this from the start.</p>
<pre class="r"><code>cp_kegg_gsea &lt;- GSEA(
  geneList = ranked_vector, 
  exponent = 1, 
  minGSSize = 10, 
  maxGSSize = 500, 
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = &quot;BH&quot;,
  TERM2GENE = kegg_term2gene, 
  TERM2NAME = kegg_term2name, 
  seed = 1234, 
  by = &quot;fgsea&quot;,
  nPermSimple = 10000
)</code></pre>
<pre><code>## using &#39;fgsea&#39; for GSEA analysis, please cite Korotkevich et al (2019).</code></pre>
<pre><code>## preparing geneSet collections...</code></pre>
<pre><code>## GSEA analysis...</code></pre>
<pre><code>## Warning in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, : There are ties in the preranked stats (0.16% of the list).
## The order of those tied genes will be arbitrary, which may produce unexpected results.</code></pre>
<pre><code>## leading edge analysis...</code></pre>
<pre><code>## done...</code></pre>
<pre class="r"><code>cp_kegg_gsea</code></pre>
<pre><code>## #
## # Gene Set Enrichment Analysis
## #
## #...@organism     UNKNOWN 
## #...@setType      UNKNOWN 
## #...@geneList     Named num [1:24419] 10.6 10.58 10.28 10.07 9.79 ...
##  - attr(*, &quot;names&quot;)= chr [1:24419] &quot;AMEX60DD020778&quot; &quot;AMEX60DD020772&quot; &quot;AMEX60DD020773&quot; &quot;AMEX60DD005693&quot; ...
## #...nPerm     
## #...pvalues adjusted by &#39;BH&#39; with cutoff &lt;0.05 
## #...29 enriched terms found
## &#39;data.frame&#39;:    29 obs. of  11 variables:
##  $ ID             : chr  &quot;map04260&quot; &quot;map04530&quot; &quot;map00601&quot; &quot;map05414&quot; ...
##  $ Description    : chr  &quot;Cardiac muscle contraction&quot; &quot;Tight junction&quot; &quot;Glycosphingolipid biosynthesis - lacto and neolacto series&quot; &quot;Dilated cardiomyopathy&quot; ...
##  $ setSize        : int  94 262 52 123 37 32 91 180 163 112 ...
##  $ enrichmentScore: num  0.752 0.639 0.807 0.685 0.826 ...
##  $ NES            : num  1.82 1.6 1.86 1.68 1.84 ...
##  $ pvalue         : num  1.50e-08 1.27e-08 2.26e-07 1.62e-06 4.94e-06 ...
##  $ p.adjust       : num  2.60e-06 2.60e-06 2.60e-05 1.40e-04 3.41e-04 ...
##  $ qvalue         : num  2.08e-06 2.08e-06 2.09e-05 1.12e-04 2.74e-04 ...
##  $ rank           : num  2313 3341 2502 3172 727 ...
##  $ leading_edge   : chr  &quot;tags=29%, list=9%, signal=26%&quot; &quot;tags=23%, list=14%, signal=20%&quot; &quot;tags=44%, list=10%, signal=40%&quot; &quot;tags=35%, list=13%, signal=31%&quot; ...
##  $ core_enrichment: chr  &quot;AMEX60DD021112/AMEX60DD012164/AMEX60DD008613/AMEX60DD021111/AMEX60DD025304/AMEX60DD054502/AMEX60DD004521/AMEX60&quot;| __truncated__ &quot;AMEX60DD021112/AMEX60DD055382/AMEX60DD012164/AMEX60DD021111/AMEX60DD025304/AMEX60DD048585/AMEX60DD054502/AMEX60&quot;| __truncated__ &quot;AMEX60DD004640/AMEX60DD013837/AMEX60DD004647/AMEX60DD004644/AMEX60DD041117/AMEX60DD004651/AMEX60DD012715/AMEX60&quot;| __truncated__ &quot;AMEX60DD012164/AMEX60DD009362/AMEX60DD008613/AMEX60DD009360/AMEX60DD025304/AMEX60DD009431/AMEX60DD054502/AMEX60&quot;| __truncated__ ...
## #...Citation
## S Xu, E Hu, Y Cai, Z Xie, X Luo, L Zhan, W Tang, Q Wang, B Liu, R Wang, W Xie, T Wu, L Xie, G Yu. Using clusterProfiler to characterize multiomics data. Nature Protocols. 2024, doi:10.1038/s41596-024-01020-z</code></pre>
<p>29 enriched terms.</p>
<p>Let’s treeplot!</p>
<pre class="r"><code># calculate pairwise similarities between the enriched terms
cp_kegg_gsea &lt;- enrichplot::pairwise_termsim(cp_kegg_gsea)
p&lt;- enrichplot::treeplot(cp_kegg_gsea, 
  showCategory = 15, 
  color = &quot;p.adjust&quot;, 
  cluster.params = list(label_words_n = 5)
)

# Add annotations (number of input genes and number of input genes with GO terms)
# Use background since all genes for ranked are in background
num_genes &lt;- length(background)
genes_with_kegg_terms &lt;- sum(background %in% kegg_term2gene$gene)

# Print the plot with custom sub-title
p &lt;- p + ggtitle(&quot;clusterProfiler GSEA of KEGG Pathways&quot;) + labs(subtitle = paste(&quot;Input genes:&quot;, num_genes, &quot;| Input genes with KEGG pathway terms:&quot;, genes_with_kegg_terms))
print(p)</code></pre>
<p><img src="novel_species_files/figure-html/treeplot%20CP%20KEGG%20GSEA-1.png" width="1152" /></p>
<p>Some muscle stuff, some cull junction stuff, and some
infection-related terms. This can be common in FEA, many genes involved
in infection responses are also part of broader stress response
pathways. These genes may be activated under different conditions, such
as environmental stress, tissue injury, or other disruptions to
homeostasis, which are common in various types of experiments. Pathways
related to immune responses can also be interconnected with pathways
controlling inflammation, wound healing, and metabolic processes. As a
result, infection-related pathways can appear in enrichment analysis
even when the experimental conditions don’t directly involve infection.
This does not mean the result is spurious - it just requires that you
exercise pragmatism, employ a basic understanding of the statistical
approach, and commit to interpreting the results in the context of your
experiment. Remember that the FEA results are to bring a large list of
genes down to a high level overview to help guide further investigation
rather than give a clear answer to your experiment.</p>
<p>I favour a volcano plot for GSEA, so we can see positive vs negative
NES. This is part of <code>ggplot</code>, not <code>enrichplot</code>,
where the <code>volplot</code> is only for ORA.</p>
<pre class="r"><code>p&lt;- ggplot(cp_kegg_gsea@result, aes(x = enrichmentScore, y = -log10(p.adjust), color = p.adjust)) + 
  geom_point(alpha = 0.7, size = 2) +  # Adjust point size
  scale_color_gradient(low = &quot;blue&quot;, high = &quot;red&quot;) +  # Color by p.adjust values
  theme_minimal() + 
  labs(title = &quot;clusterProfiler GSEA of KEGG Pathways&quot;, 
       x = &quot;Enrichment Score (NES)&quot;, 
       y = &quot;-log10(Adjusted P-value)&quot;,
       color = &quot;Adjusted P-value&quot;) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels for readability
  geom_vline(xintercept = 0, linetype = &quot;dashed&quot;, color = &quot;black&quot;) +  # Add vertical line at x=0
  geom_hline(yintercept = -log10(0.05), linetype = &quot;dashed&quot;, color = &quot;black&quot;) +  # Add horizontal line at p=0.05 cutoff
  geom_text(aes(label = Description), 
            hjust = 0.5, 
            vjust = -0.5,  # Move labels higher off the points
            size = 3, 
            check_overlap = TRUE, 
            alpha = 0.7)  # Add labels for each pathway term

print(p)</code></pre>
<p><img src="novel_species_files/figure-html/volcano%20plot%20CP%20gsea%20kegg,%20-1.png" width="864" /></p>
<p>Interesting that all terms except one have leading edge genes that
are upregulated in distal compared to proximal (the reference
level)!</p>
</div>
</div>
<div id="reformat-annotation-files-for-webgestaltr-go-and-kegg-analysis"
class="section level1">
<h1>5. Reformat annotation files for <code>WebGestaltR</code> GO and
KEGG analysis</h1>
<p>GMT files must have <code>.gmt</code> suffix and description files
must have <code>.des</code> suffix.</p>
<div id="create-gmt-objects" class="section level2">
<h2>5.1 Create GMT objects</h2>
<p>The GMT files need links for all of the terms, so that we can have
that handy link-out to enriched terms from the HTML report we
experienced in the last activity. This is actually pretty simple to do
thanks to consistent URLs.</p>
<p>For GO, we just need to paste the term ID to the end of this link <a
href="https://www.ebi.ac.uk/QuickGO/term/"
class="uri">https://www.ebi.ac.uk/QuickGO/term/</a><GOID></p>
<p>And for KEGG, we need to paste the map ID to the end of this link: <a
href="https://www.genome.jp/dbget-bin/www_bget"
class="uri">https://www.genome.jp/dbget-bin/www_bget</a>?<mapID></p>
<div id="go-gmt" class="section level3">
<h3>5.1.1 GO GMT</h3>
<p>Note that in the below code, the first command is identical the one
that created the <code>go_term2gene</code> object earlier in the
notebook. we could just use the <code>go_term2gene</code> object and
skip step of the below code, using <code>go_term2gene</code> as input
for step 2 rather than <code>go_data</code>. This code duplication is
intentional, so that this code chunk is standalone for re-use and
re-purpose.</p>
<pre class="r"><code># Step 1: Extract relevant columns (GO terms and gene IDs) from eggnog_anno
go_data &lt;- eggnog_anno %&gt;%
  dplyr::select(GOs, `#query`) %&gt;%  # Select the GO terms and the gene IDs
  dplyr::filter(GOs != &quot;-&quot;) %&gt;%  # Filter out rows where the GO ID is missing (&quot;-&quot;)
  separate_rows(GOs, sep = &quot;,&quot;) %&gt;%  # Split comma-delimited list of GO terms into separate rows
  dplyr::select(GOs, `#query`) %&gt;%  # Keep GO terms and gene IDs columns
  distinct() %&gt;%  # Remove duplicates
  drop_na()  # Drop any rows with missing values

# Rename columns to match the format (term, gene)
colnames(go_data) &lt;- c(&quot;term&quot;, &quot;gene&quot;)

# Step 2: Create external links for each GO term (link to QuickGO)
go_data &lt;- go_data %&gt;%
  dplyr::mutate(external_link = paste0(&quot;https://www.ebi.ac.uk/QuickGO/term/&quot;, term))

# Step 3: Group genes by GO term and concatenate gene list by tab
go_term_grouped &lt;- go_data %&gt;%
  dplyr::group_by(term) %&gt;%
  dplyr::summarize(genes = paste(gene, collapse = &quot;\t&quot;), .groups = &quot;drop&quot;)

# Step 4: Add the external link for each GO term
go_term_grouped &lt;- go_term_grouped %&gt;%
  dplyr::left_join(go_data %&gt;% dplyr::select(term, external_link) %&gt;% distinct(), by = &quot;term&quot;)

# Step 5: Create the final GMT format entry (term ID, external link, and gene list)
go_gmt &lt;- go_term_grouped %&gt;%
  dplyr::mutate(gmt_entry = paste(term, external_link, genes, sep = &quot;\t&quot;)) %&gt;%
  dplyr::select(gmt_entry)

# Save to file
write.table(go_gmt, file = &quot;Axolotl_GO.gmt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = FALSE)</code></pre>
</div>
<div id="kegg-gmt" class="section level3">
<h3>5.1.2 KEGG GMT</h3>
<p>As above, for clarity we have avoided using the
<code>TERM2GENE</code> object to ensure this code chunk can be
standalone.</p>
<pre class="r"><code># Step 1: Extract relevant columns (KEGG Pathway and gene IDs) from eggnog_anno
kegg_data &lt;- eggnog_anno %&gt;%
  dplyr::select(KEGG_Pathway, `#query`) %&gt;%  # Select the KEGG Pathway and gene ID columns
  dplyr::filter(grepl(&quot;map&quot;, KEGG_Pathway)) %&gt;%  # Keep only rows where KEGG_Pathway contains &#39;map&#39;
  separate_rows(KEGG_Pathway, sep = &quot;,&quot;) %&gt;%  # Split multiple pathways into separate rows
  dplyr::mutate(term = gsub(&quot;map:&quot;, &quot;&quot;, KEGG_Pathway)) %&gt;%  # Remove the &quot;map:&quot; prefix
  dplyr::filter(grepl(&quot;^map&quot;, term)) %&gt;%  # Filter again to keep only &#39;map&#39; pathways (after removing &quot;map:&quot;)
  dplyr::select(term, `#query`) %&gt;%  # Select the KEGG Pathway and gene ID columns
  distinct() %&gt;%  # Remove duplicate rows
  drop_na()  # Remove rows with missing values

# Ensure the column is properly named
colnames(kegg_data)[colnames(kegg_data) == &quot;#query&quot;] &lt;- &quot;gene&quot;

# Step 2: Create external links for each KEGG pathway
kegg_data &lt;- kegg_data %&gt;%
  dplyr::mutate(external_link = paste0(&quot;https://www.genome.jp/dbget-bin/www_bget?&quot;, term))

# Step 3: Group by KEGG pathway term and concatenate the gene list
kegg_term_grouped &lt;- kegg_data %&gt;%
  dplyr::group_by(term) %&gt;%
  dplyr::summarize(genes = paste(gene, collapse = &quot;\t&quot;), .groups = &quot;drop&quot;)

# Step 4: Add the external link for each KEGG pathway
kegg_term_grouped &lt;- kegg_term_grouped %&gt;%
  dplyr::left_join(kegg_data %&gt;% dplyr::select(term, external_link) %&gt;% distinct(), by = &quot;term&quot;)

# Step 5: Create the final GMT format entry (Pathway, External Link, Genes)
kegg_gmt &lt;- kegg_term_grouped %&gt;%
  dplyr::mutate(gmt_entry = paste(term, external_link, genes, sep = &quot;\t&quot;)) %&gt;%
  dplyr::select(gmt_entry)

# Save to file
write.table(kegg_gmt, file = &quot;Axolotl_KEGG-pathways.gmt&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = FALSE)</code></pre>
</div>
</div>
<div id="create-description-objects" class="section level2">
<h2>5.2 Create description objects</h2>
<div id="go-description" class="section level3">
<h3>5.2.1 GO description</h3>
<p>The description file is identical to <code>clusterProfiler</code>
<code>TERM2NAME</code> . Again, we don’t want to use any code from the
creation of <code>clusterProfiler</code> objects to ensure this part can
be used alone.</p>
<pre class="r"><code># Step 1: Get terms from ontology object created from go.obo with ontologyIndex function: 
ontology_term_names &lt;- ontology$name 

# Step 2: Filter and separate GO terms from the annotation file
# We filter out rows where no GO terms are assigned and separate comma-delimited GO terms
go_terms &lt;- eggnog_anno %&gt;%
  dplyr::select(GOs, `#query`) %&gt;%
  dplyr::filter(GOs != &quot;-&quot;) %&gt;%  # Keep only rows with GO terms
  separate_rows(GOs, sep = &quot;,&quot;) %&gt;%
  dplyr::mutate(term = GOs) %&gt;%
  dplyr::select(term, `#query`) %&gt;%
  distinct() %&gt;%
  drop_na()  # Drop rows with missing values

# Step 3: Create the description by matching GO terms to their names in the ontology
go_des &lt;- go_terms %&gt;%
  dplyr::mutate(name = ontology_term_names[term]) %&gt;%  # Map term to its name from the ontology
  dplyr::select(term, name) %&gt;%  # Keep only the term and name
  distinct() %&gt;%  # Remove duplicates
  drop_na() %&gt;%  # Remove rows with missing values
  filter(!grepl(&quot;obsolete&quot;, name))  # Remove obsolete terms if present

# Save to file
write.table(go_des, file = &quot;Axolotl_GO.des&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = FALSE)

# Check
head(go_des)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"GO:0000981","2":"DNA-binding transcription factor activity, RNA polymerase II-specific"},{"1":"GO:0003674","2":"molecular_function"},{"1":"GO:0003700","2":"DNA-binding transcription factor activity"},{"1":"GO:0006355","2":"regulation of DNA-templated transcription"},{"1":"GO:0006357","2":"regulation of transcription by RNA polymerase II"},{"1":"GO:0008150","2":"biological_process"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="kegg-description-file" class="section level3">
<h3>5.2.2 KEGG description file</h3>
<p>More extra work for the sake of portability :-)</p>
<pre class="r"><code># Get columns from kegg gmt object 
kegg_gmt_columns &lt;- kegg_gmt %&gt;%
  separate(gmt_entry, into = c(&quot;term&quot;, &quot;external_link&quot;, &quot;genes&quot;), sep = &quot;\t&quot;)</code></pre>
<pre><code>## Warning: Expected 3 pieces. Additional pieces discarded in 381 rows [1, 2, 3, 4, 5, 6,
## 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...].</code></pre>
<pre class="r"><code># Create the kegg_des table by joining the pathways file with the species-specific terms from kegg_gmt
kegg_des &lt;- kegg_pathways %&gt;%
  dplyr::filter(term %in% kegg_gmt_columns$term) %&gt;%
  dplyr::select(term, name) %&gt;%
  distinct() %&gt;%
  drop_na()

# Save to file
write.table(kegg_des, file = &quot;Axolotl_KEGG-pathways.des&quot;, sep = &quot;\t&quot;, quote = FALSE, row.names = FALSE, col.names = FALSE)

# Check the first few rows
head(kegg_des)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["term"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"map01100","2":"Metabolic pathways","_rn_":"1"},{"1":"map01110","2":"Biosynthesis of secondary metabolites","_rn_":"2"},{"1":"map01120","2":"Microbial metabolism in diverse environments","_rn_":"3"},{"1":"map01200","2":"Carbon metabolism","_rn_":"4"},{"1":"map01210","2":"2-Oxocarboxylic acid metabolism","_rn_":"5"},{"1":"map01212","2":"Fatty acid metabolism","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</div>
<div id="run-webgestaltr-ora-and-gsea-with-custom-database-files"
class="section level1">
<h1>6. Run <code>WebGestaltR</code> ORA and GSEA with custom database
files</h1>
<p>We will use the same approach as with <code>clusterProfiler</code>
and run ORA with GO and GSEA with KEGG for both tools.</p>
<div id="webgestaltr-ora-of-go-terms" class="section level2">
<h2>6.1 WebGestaltR ORA of GO terms</h2>
<pre class="r"><code>outputDirectory &lt;- &quot;WebGestaltR_results&quot; 
project &lt;- &quot;Axolotl_ORA_GO&quot;

WebGestaltR(
    organism = &quot;others&quot;,                          # must specify &#39;others&#39; when using custom db files
    enrichMethod = &quot;ORA&quot;,                         # Perform ORA, GSEA or NTA
    interestGene = degs,                          # gene list of interest
    referenceGene = background,                   # background genes
    enrichDatabaseFile = &quot;Axolotl_GO.gmt&quot;,        # the custom gmt file
    enrichDatabaseDescriptionFile = &quot;Axolotl_GO.des&quot;,   # the custom description file  
    isOutput = TRUE,                              # Set to FALSE if you don&#39;t want files saved to disk
    fdrMethod = &quot;BH&quot;,                             # Benjamini-Hochberg multiple testing correction
    sigMethod = &quot;fdr&quot;,                            # Significance method (&#39;fdr&#39; or &#39;top&#39;)
    fdrThr = 0.05,                                # FDR significance threshold
    minNum = 10,                                   # Minimum number of genes per category
    maxNum = 500,                                 # Maximum number of genes per category
    boxplot = TRUE,
    outputDirectory = outputDirectory,
    projectName = project
)</code></pre>
<pre><code>## Loading the functional categories...
## Loading the ID list...
## Loading the reference list...</code></pre>
<pre><code>## Warning in dir.create(projectDir): &#39;WebGestaltR_results/Project_Axolotl_ORA_GO&#39;
## already exists</code></pre>
<pre><code>## Performing the enrichment analysis...
## Begin affinity propagation...
## End affinity propagation...
## Begin weighted set cover...
## End weighted set cover...
## Generate the final report...
## Results can be found in the WebGestaltR_results/Project_Axolotl_ORA_GO!</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["geneSet"],"name":[1],"type":["chr"],"align":["left"]},{"label":["description"],"name":[2],"type":["chr"],"align":["left"]},{"label":["link"],"name":[3],"type":["chr"],"align":["left"]},{"label":["size"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["overlap"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["expect"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["enrichmentRatio"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["pValue"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["FDR"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["overlapId"],"name":[10],"type":["chr"],"align":["left"]}],"data":[{"1":"GO:0048821","2":"erythrocyte development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0048821","4":"45","5":"7","6":"0.43292197","7":"16.169195","8":"2.204721e-07","9":"0.0008909043","10":"AMEX60DD025537;AMEX60DD026264;AMEX60DD026267;AMEX60DD032898;AMEX60DD032958;AMEX60DD032960;AMEX60DD002868"},{"1":"GO:0070268","2":"cornification","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0070268","4":"46","5":"7","6":"0.44254246","7":"15.817691","8":"2.579655e-07","9":"0.0008909043","10":"AMEX60DD016775;AMEX60DD038318;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004554;AMEX60DD010118"},{"1":"GO:0031424","2":"keratinization","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0031424","4":"48","5":"7","6":"0.46178344","7":"15.158621","8":"3.492104e-07","9":"0.0008909043","10":"AMEX60DD016775;AMEX60DD038318;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004554;AMEX60DD010118"},{"1":"GO:0019317","2":"fucose catabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0019317","4":"18","5":"5","6":"0.17316879","7":"28.873563","8":"5.957234e-07","9":"0.0008909043","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0042354","2":"L-fucose metabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0042354","4":"18","5":"5","6":"0.17316879","7":"28.873563","8":"5.957234e-07","9":"0.0008909043","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0042355","2":"L-fucose catabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0042355","4":"18","5":"5","6":"0.17316879","7":"28.873563","8":"5.957234e-07","9":"0.0008909043","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0008544","2":"epidermis development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0008544","4":"379","5":"16","6":"3.64616507","7":"4.388172","8":"7.520091e-07","9":"0.0009639682","10":"AMEX60DD016775;AMEX60DD018597;AMEX60DD025054;AMEX60DD029166;AMEX60DD038224;AMEX60DD038318;AMEX60DD039516;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004414;AMEX60DD004554;AMEX60DD005188;AMEX60DD007528;AMEX60DD008951;AMEX60DD010118"},{"1":"GO:0030057","2":"desmosome","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0030057","4":"20","5":"5","6":"0.19240977","7":"25.986207","8":"1.061384e-06","9":"0.0011376829","10":"AMEX60DD038318;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004554"},{"1":"GO:0001533","2":"cornified envelope","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0001533","4":"21","5":"5","6":"0.20203025","7":"24.748768","8":"1.382307e-06","9":"0.0011376829","10":"AMEX60DD038318;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004554"},{"1":"GO:0003229","2":"ventricular cardiac muscle tissue development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0003229","4":"59","5":"7","6":"0.56760881","7":"12.332437","8":"1.481389e-06","9":"0.0011376829","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD039606;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0060415","2":"muscle tissue morphogenesis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0060415","4":"85","5":"8","6":"0.81774151","7":"9.783043","8":"1.557569e-06","9":"0.0011376829","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD000147;AMEX60DD001828;AMEX60DDU001014191"},{"1":"GO:0008417","2":"fucosyltransferase activity","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0008417","4":"22","5":"5","6":"0.21165074","7":"23.623824","8":"1.775054e-06","9":"0.0011376829","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0030506","2":"ankyrin binding","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0030506","4":"22","5":"5","6":"0.21165074","7":"23.623824","8":"1.775054e-06","9":"0.0011376829","10":"AMEX60DD023613;AMEX60DD032898;AMEX60DD004168;AMEX60DD009895;AMEX60DD012067"},{"1":"GO:0042415","2":"norepinephrine metabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0042415","4":"22","5":"5","6":"0.21165074","7":"23.623824","8":"1.775054e-06","9":"0.0011376829","10":"AMEX60DD026518;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0043588","2":"skin development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0043588","4":"228","5":"12","6":"2.19347134","7":"5.470780","8":"2.068300e-06","9":"0.0012372569","10":"AMEX60DD016775;AMEX60DD025054;AMEX60DD030862;AMEX60DD038318;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004554;AMEX60DD005188;AMEX60DD007528;AMEX60DD010118;AMEX60DD010980"},{"1":"GO:0046920","2":"alpha-(1->3)-fucosyltransferase activity","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0046920","4":"11","5":"4","6":"0.10582537","7":"37.798119","8":"2.573549e-06","9":"0.0014432782","10":"AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0050432","2":"catecholamine secretion","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0050432","4":"12","5":"4","6":"0.11544586","7":"34.648276","8":"3.831477e-06","9":"0.0020223437","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0006004","2":"fucose metabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0006004","4":"26","5":"5","6":"0.25013270","7":"19.989390","8":"4.298605e-06","9":"0.0020300728","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0140103","2":"catalytic activity, acting on a glycoprotein","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0140103","4":"26","5":"5","6":"0.25013270","7":"19.989390","8":"4.298605e-06","9":"0.0020300728","10":"AMEX60DD013837;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0030548","2":"acetylcholine receptor regulator activity","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0030548","4":"13","5":"4","6":"0.12506635","7":"31.983024","8":"5.493012e-06","9":"0.0023470854","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0099602","2":"neurotransmitter receptor regulator activity","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0099602","4":"13","5":"4","6":"0.12506635","7":"31.983024","8":"5.493012e-06","9":"0.0023470854","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0055010","2":"ventricular cardiac muscle tissue morphogenesis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0055010","4":"50","5":"6","6":"0.48102442","7":"12.473379","8":"8.015007e-06","9":"0.0032690299","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0036065","2":"fucosylation","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0036065","4":"30","5":"5","6":"0.28861465","7":"17.324138","8":"9.028492e-06","9":"0.0035222895","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0048644","2":"muscle organ morphogenesis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0048644","4":"108","5":"8","6":"1.03901274","7":"7.699617","8":"9.460891e-06","9":"0.0035371906","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD000147;AMEX60DD001828;AMEX60DDU001014191"},{"1":"GO:0061515","2":"myeloid cell development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0061515","4":"81","5":"7","6":"0.77925955","7":"8.982886","8":"1.265999e-05","9":"0.0045439229","10":"AMEX60DD025537;AMEX60DD026264;AMEX60DD026267;AMEX60DD032898;AMEX60DD032958;AMEX60DD032960;AMEX60DD002868"},{"1":"GO:0015669","2":"gas transport","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0015669","4":"18","5":"4","6":"0.17316879","7":"23.098851","8":"2.264416e-05","9":"0.0078148475","10":"AMEX60DD021527;AMEX60DD026264;AMEX60DD026267;AMEX60DD032898"},{"1":"GO:0006584","2":"catecholamine metabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0006584","4":"61","5":"6","6":"0.58684979","7":"10.224081","8":"2.567666e-05","9":"0.0082284518","10":"AMEX60DD015068;AMEX60DD026518;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0009712","2":"catechol-containing compound metabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0009712","4":"61","5":"6","6":"0.58684979","7":"10.224081","8":"2.567666e-05","9":"0.0082284518","10":"AMEX60DD015068;AMEX60DD026518;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0098609","2":"cell-cell adhesion","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0098609","4":"394","5":"14","6":"3.79047240","7":"3.693471","8":"2.693982e-05","9":"0.0083355525","10":"AMEX60DD023613;AMEX60DD029929;AMEX60DD030102;AMEX60DD035815;AMEX60DD038318;AMEX60DD039516;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD047520;AMEX60DD002483;AMEX60DD005631;AMEX60DD007528;AMEX60DD007977"},{"1":"GO:0030216","2":"keratinocyte differentiation","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0030216","4":"93","5":"7","6":"0.89470541","7":"7.823804","8":"3.133565e-05","9":"0.0093724919","10":"AMEX60DD016775;AMEX60DD038318;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD004554;AMEX60DD010118"},{"1":"GO:0033130","2":"acetylcholine receptor binding","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0033130","4":"20","5":"4","6":"0.19240977","7":"20.788966","8":"3.532049e-05","9":"0.0102235737","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0003206","2":"cardiac chamber morphogenesis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0003206","4":"130","5":"8","6":"1.25066348","7":"6.396605","8":"3.654950e-05","9":"0.0102487095","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD000147;AMEX60DD001828;AMEX60DDU001014191"},{"1":"GO:0055008","2":"cardiac muscle tissue morphogenesis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0055008","4":"68","5":"6","6":"0.65419321","7":"9.171602","8":"4.789645e-05","9":"0.0130234813","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0051937","2":"catecholamine transport","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0051937","4":"23","5":"4","6":"0.22127123","7":"18.077361","8":"6.312123e-05","9":"0.0166584354","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0003208","2":"cardiac ventricle morphogenesis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0003208","4":"74","5":"6","6":"0.71191614","7":"8.427959","8":"7.732207e-05","9":"0.0198231686","10":"AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0003205","2":"cardiac chamber development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0003205","4":"188","5":"9","6":"1.80865180","7":"4.976082","8":"8.520438e-05","9":"0.0212371914","10":"AMEX60DD021953;AMEX60DD030069;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD000147;AMEX60DD001828;AMEX60DDU001014191"},{"1":"GO:0086080","2":"protein binding involved in heterotypic cell-cell adhesion","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0086080","4":"10","5":"3","6":"0.09620488","7":"31.183448","8":"9.960638e-05","9":"0.0241558924","10":"AMEX60DD038318;AMEX60DD039599;AMEX60DD039606"},{"1":"GO:0098742","2":"cell-cell adhesion via plasma-membrane adhesion molecules","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0098742","4":"154","5":"8","6":"1.48155520","7":"5.399731","8":"1.209542e-04","9":"0.0279353861","10":"AMEX60DD030102;AMEX60DD035815;AMEX60DD039599;AMEX60DD039606;AMEX60DD047520;AMEX60DD002483;AMEX60DD005631;AMEX60DD007528"},{"1":"GO:0015701","2":"bicarbonate transport","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0015701","4":"27","5":"4","6":"0.25975318","7":"15.399234","8":"1.214176e-04","9":"0.0279353861","10":"AMEX60DD026264;AMEX60DD026267;AMEX60DD032898;AMEX60DD009895"},{"1":"GO:0002934","2":"desmosome organization","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0002934","4":"11","5":"3","6":"0.10582537","7":"28.348589","8":"1.359937e-04","9":"0.0290540799","10":"AMEX60DD038318;AMEX60DD039606;AMEX60DD004554"},{"1":"GO:0008343","2":"adult feeding behavior","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0008343","4":"11","5":"3","6":"0.10582537","7":"28.348589","8":"1.359937e-04","9":"0.0290540799","10":"AMEX60DD025054;AMEX60DD005942;AMEX60DD007977"},{"1":"GO:0086042","2":"cardiac muscle cell-cardiac muscle cell adhesion","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0086042","4":"11","5":"3","6":"0.10582537","7":"28.348589","8":"1.359937e-04","9":"0.0290540799","10":"AMEX60DD038318;AMEX60DD039599;AMEX60DD039606"},{"1":"GO:0009312","2":"oligosaccharide biosynthetic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0009312","4":"28","5":"4","6":"0.26937367","7":"14.849261","8":"1.405999e-04","9":"0.0293396037","10":"AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0070161","2":"anchoring junction","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0070161","4":"407","5":"13","6":"3.91553875","7":"3.320105","8":"1.540264e-04","9":"0.0302619542","10":"AMEX60DD014166;AMEX60DD023613;AMEX60DD030102;AMEX60DD038318;AMEX60DD039516;AMEX60DD039599;AMEX60DD039603;AMEX60DD039606;AMEX60DD047520;AMEX60DD055791;AMEX60DD004554;AMEX60DD010314;AMEX60DD010980"},{"1":"GO:0095500","2":"acetylcholine receptor signaling pathway","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0095500","4":"29","5":"4","6":"0.27899416","7":"14.337218","8":"1.618827e-04","9":"0.0302619542","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:1903831","2":"signal transduction involved in cellular response to ammonium ion","3":"https://www.ebi.ac.uk/QuickGO/term/GO:1903831","4":"29","5":"4","6":"0.27899416","7":"14.337218","8":"1.618827e-04","9":"0.0302619542","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:1905144","2":"response to acetylcholine","3":"https://www.ebi.ac.uk/QuickGO/term/GO:1905144","4":"29","5":"4","6":"0.27899416","7":"14.337218","8":"1.618827e-04","9":"0.0302619542","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:1905145","2":"cellular response to acetylcholine","3":"https://www.ebi.ac.uk/QuickGO/term/GO:1905145","4":"29","5":"4","6":"0.27899416","7":"14.337218","8":"1.618827e-04","9":"0.0302619542","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0032580","2":"Golgi cisterna membrane","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0032580","4":"54","5":"5","6":"0.51950637","7":"9.624521","8":"1.664593e-04","9":"0.0304824345","10":"AMEX60DD017241;AMEX60DD004640;AMEX60DD004644;AMEX60DD004647;AMEX60DD004651"},{"1":"GO:0005833","2":"hemoglobin complex","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0005833","4":"12","5":"3","6":"0.11544586","7":"25.986207","8":"1.800478e-04","9":"0.0323113706","10":"AMEX60DD021527;AMEX60DD026264;AMEX60DD026267"},{"1":"GO:0030218","2":"erythrocyte differentiation","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0030218","4":"124","5":"7","6":"1.19294055","7":"5.867853","8":"1.944082e-04","9":"0.0342044155","10":"AMEX60DD025537;AMEX60DD026264;AMEX60DD026267;AMEX60DD032898;AMEX60DD032958;AMEX60DD032960;AMEX60DD002868"},{"1":"GO:0018958","2":"phenol-containing compound metabolic process","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0018958","4":"166","5":"8","6":"1.59700106","7":"5.009389","8":"2.028930e-04","9":"0.0350107524","10":"AMEX60DD015068;AMEX60DD023613;AMEX60DD026518;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001001189;AMEX60DDU001014191"},{"1":"GO:0098631","2":"cell adhesion mediator activity","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0098631","4":"31","5":"4","6":"0.29823514","7":"13.412236","8":"2.112828e-04","9":"0.0357705733","10":"AMEX60DD018132;AMEX60DD038318;AMEX60DD039599;AMEX60DD039606"},{"1":"GO:0015844","2":"monoamine transport","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0015844","4":"32","5":"4","6":"0.30785563","7":"12.993103","8":"2.396714e-04","9":"0.0391013038","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0030547","2":"signaling receptor inhibitor activity","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0030547","4":"32","5":"4","6":"0.30785563","7":"12.993103","8":"2.396714e-04","9":"0.0391013038","10":"AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DDU001014191"},{"1":"GO:0015850","2":"organic hydroxy compound transport","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0015850","4":"130","5":"7","6":"1.25066348","7":"5.597029","8":"2.600876e-04","9":"0.0416743871","10":"AMEX60DD021914;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD044012;AMEX60DD004446;AMEX60DDU001014191"},{"1":"GO:0034101","2":"erythrocyte homeostasis","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0034101","4":"132","5":"7","6":"1.26990446","7":"5.512226","8":"2.855717e-04","9":"0.0441799172","10":"AMEX60DD025537;AMEX60DD026264;AMEX60DD026267;AMEX60DD032898;AMEX60DD032958;AMEX60DD032960;AMEX60DD002868"},{"1":"GO:0035265","2":"organ growth","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0035265","4":"132","5":"7","6":"1.26990446","7":"5.512226","8":"2.855717e-04","9":"0.0441799172","10":"AMEX60DD014166;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD046037;AMEX60DD007977;AMEX60DDU001014191"},{"1":"GO:0015671","2":"oxygen transport","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0015671","4":"14","5":"3","6":"0.13468684","7":"22.273892","8":"2.937182e-04","9":"0.0446700577","10":"AMEX60DD021527;AMEX60DD026264;AMEX60DD026267"},{"1":"GO:0014704","2":"intercalated disc","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0014704","4":"62","5":"5","6":"0.59647028","7":"8.382647","8":"3.202523e-04","9":"0.0478937319","10":"AMEX60DD021112;AMEX60DD038318;AMEX60DD039599;AMEX60DD039606;AMEX60DD009895"},{"1":"GO:0007517","2":"muscle organ development","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0007517","4":"440","5":"13","6":"4.23301486","7":"3.071097","8":"3.285646e-04","9":"0.0483313073","10":"AMEX60DD014166;AMEX60DD029929;AMEX60DD030069;AMEX60DD037141;AMEX60DD038318;AMEX60DD040658;AMEX60DD041071;AMEX60DD041072;AMEX60DD046037;AMEX60DD000147;AMEX60DD001828;AMEX60DD005188;AMEX60DDU001014191"},{"1":"GO:0031581","2":"hemidesmosome assembly","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0031581","4":"35","5":"4","6":"0.33671709","7":"11.879409","8":"3.412575e-04","9":"0.0487905703","10":"AMEX60DD018597;AMEX60DD030862;AMEX60DD039516;AMEX60DD008951"},{"1":"GO:0005913","2":"","3":"https://www.ebi.ac.uk/QuickGO/term/GO:0005913","4":"136","5":"7","6":"1.30838641","7":"5.350101","8":"3.425617e-04","9":"0.0487905703","10":"AMEX60DD023613;AMEX60DD030102;AMEX60DD038318;AMEX60DD039516;AMEX60DD039599;AMEX60DD047520;AMEX60DD004554"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Open the HTML report
<code>WebGestaltR_results/Project_Axolotl_ORA_GO/Report_Axolotl_ORA_GO.html</code>
from the files pane in a browser.</p>
<p>Note some term similarity to what we have seen with the past 2
analyses (that’s reassuring!)</p>
<p>Change the ‘Enrichment Results’ view from table to ‘Bar chart’, then
try the ‘Affinity propagation’ and ‘Weighted set cover’ term clustering
algorithms. Both methods help focus the results.</p>
<p>Confirm that our GMT file correctly included the term link by
selecting a term and clicking the hyperlink at <code>Analyte set</code>.
Pretty neat huh :-)</p>
</div>
<div id="webgestaltr-gsea-of-kegg-pathways" class="section level2">
<h2>6.2 WebGestaltR GSEA of KEGG Pathways</h2>
<p>This will take slightly longer than ORA. We will set threads to 7</p>
<pre class="r"><code>outputDirectory &lt;- &quot;WebGestaltR_results&quot; 
project &lt;- &quot;Axolotl_GSEA_KEGG&quot;

suppressWarnings({ WebGestaltR(
    organism = &quot;others&quot;,                          # must specify &#39;others&#39; when using custom db files
    enrichMethod = &quot;GSEA&quot;,                        # Perform ORA, GSEA or NTA
    interestGene = ranked_df,                     # ranked dataframe
    enrichDatabaseFile = &quot;Axolotl_KEGG-pathways.gmt&quot;,        # the custom gmt file
    enrichDatabaseDescriptionFile = &quot;Axolotl_KEGG-pathways.des&quot;,   # the custom description file  
    isOutput = TRUE,                              # Set to FALSE if you don&#39;t want files saved to disk
    fdrMethod = &quot;BH&quot;,                             # Benjamini-Hochberg multiple testing correction
    sigMethod = &quot;fdr&quot;,                            # Significance method (&#39;fdr&#39; or &#39;top&#39;)
    fdrThr = 0.05,                                # FDR significance threshold
    minNum = 10,                                   # Minimum number of genes per category
    maxNum = 500,                                 # Maximum number of genes per category
    outputDirectory = outputDirectory,
    projectName = project,
    nThreads = 7
) })</code></pre>
<pre><code>## Loading the functional categories...
## Loading the ID list...
## Performing the enrichment analysis...
## 1000 permutations of score complete...</code></pre>
<pre><code>## Begin affinity propagation...
## End affinity propagation...
## Begin weighted set cover...
## End weighted set cover...
## Generate the final report...
## Results can be found in the WebGestaltR_results/Project_Axolotl_GSEA_KEGG!</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["geneSet"],"name":[1],"type":["chr"],"align":["left"]},{"label":["description"],"name":[2],"type":["chr"],"align":["left"]},{"label":["link"],"name":[3],"type":["chr"],"align":["left"]},{"label":["enrichmentScore"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["normalizedEnrichmentScore"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["pValue"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["FDR"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["size"],"name":[8],"type":["int"],"align":["right"]},{"label":["plotPath"],"name":[9],"type":["chr"],"align":["left"]},{"label":["leadingEdgeNum"],"name":[10],"type":["int"],"align":["right"]},{"label":["leadingEdgeId"],"name":[11],"type":["chr"],"align":["left"]}],"data":[{"1":"map04260","2":"Cardiac muscle contraction","3":"https://www.genome.jp/dbget-bin/www_bget?map04260","4":"0.7520517","5":"1.824532","6":"0.000000000","7":"0.0004799410","8":"94","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map04260.png","10":"27","11":"AMEX60DD021112;AMEX60DD012164;AMEX60DD008613;AMEX60DD021111;AMEX60DD025304;AMEX60DD054502;AMEX60DD004521;AMEX60DD012899;AMEX60DD016179;AMEX60DD018138;AMEX60DD031139;AMEX60DD048093;AMEX60DD009525;AMEX60DD051226;AMEX60DD009714;AMEX60DD034544;AMEX60DD012123;AMEX60DD053503;AMEX60DD024402;AMEX60DDU001007100;AMEX60DD004576;AMEX60DD016515;AMEX60DD021091;AMEX60DD035296;AMEX60DD023720;AMEX60DD024110;AMEX60DD024383"},{"1":"map00515","2":"Mannose type O-glycan biosynthesis","3":"https://www.genome.jp/dbget-bin/www_bget?map00515","4":"0.8256989","5":"1.797527","6":"0.000000000","7":"0.0006399213","8":"37","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00515.png","10":"8","11":"AMEX60DD004640;AMEX60DD004647;AMEX60DD004644;AMEX60DD041117;AMEX60DD004651;AMEX60DD030769;AMEX60DD041108;AMEX60DD004650"},{"1":"map00603","2":"Glycosphingolipid biosynthesis - globo and isoglobo series","3":"https://www.genome.jp/dbget-bin/www_bget?map00603","4":"0.8226890","5":"1.772253","6":"0.000000000","7":"0.0007199115","8":"32","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00603.png","10":"16","11":"AMEX60DD004640;AMEX60DD004647;AMEX60DD004644;AMEX60DD041117;AMEX60DD004651;AMEX60DD041108;AMEX60DD004650;AMEX60DD012874;AMEX60DD017486;AMEX60DD017237;AMEX60DD017241;AMEX60DD012876;AMEX60DD042400;AMEX60DD017483;AMEX60DD017234;AMEX60DD017240"},{"1":"map00601","2":"Glycosphingolipid biosynthesis - lacto and neolacto series","3":"https://www.genome.jp/dbget-bin/www_bget?map00601","4":"0.8066736","5":"1.837734","6":"0.000000000","7":"0.0009598819","8":"52","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00601.png","10":"23","11":"AMEX60DD004640;AMEX60DD013837;AMEX60DD004647;AMEX60DD004644;AMEX60DD041117;AMEX60DD004651;AMEX60DD012715;AMEX60DD041108;AMEX60DD032040;AMEX60DD004650;AMEX60DD012874;AMEX60DD017486;AMEX60DD017237;AMEX60DD025023;AMEX60DD017241;AMEX60DD012876;AMEX60DDU001007639;AMEX60DD017483;AMEX60DD017234;AMEX60DD013845;AMEX60DD017240;AMEX60DD014059;AMEX60DD012980"},{"1":"map00450","2":"Selenocompound metabolism","3":"https://www.genome.jp/dbget-bin/www_bget?map00450","4":"0.7653822","5":"1.678501","6":"0.000000000","7":"0.0075190752","8":"35","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00450.png","10":"10","11":"AMEX60DD055022;AMEX60DD053800;AMEX60DD053670;AMEX60DD056182;AMEX60DD008432;AMEX60DD008434;AMEX60DD038409;AMEX60DD053647;AMEX60DD029119;AMEX60DD038408"},{"1":"map05414","2":"Dilated cardiomyopathy","3":"https://www.genome.jp/dbget-bin/www_bget?map05414","4":"0.6850575","5":"1.681587","6":"0.000000000","7":"0.0076790555","8":"123","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05414.png","10":"43","11":"AMEX60DD012164;AMEX60DD009362;AMEX60DD008613;AMEX60DD009360;AMEX60DD025304;AMEX60DD009431;AMEX60DD054502;AMEX60DD009364;AMEX60DD004521;AMEX60DD005313;AMEX60DD031139;AMEX60DD055791;AMEX60DD015201;AMEX60DD025721;AMEX60DD030862;AMEX60DD034544;AMEX60DD009353;AMEX60DD056342;AMEX60DD012123;AMEX60DD024402;AMEX60DD009365;AMEX60DDU001007100;AMEX60DD009388;AMEX60DD010588;AMEX60DD009354;AMEX60DD035296;AMEX60DD042070;AMEX60DD009940;AMEX60DD009427;AMEX60DD023720;AMEX60DD024110;AMEX60DD049332;AMEX60DD009395;AMEX60DD010587;AMEX60DD025328;AMEX60DD017438;AMEX60DD032746;AMEX60DD009358;AMEX60DD009400;AMEX60DD055675;AMEX60DD009396;AMEX60DD048456;AMEX60DD006925"},{"1":"map05412","2":"Arrhythmogenic right ventricular cardiomyopathy","3":"https://www.genome.jp/dbget-bin/www_bget?map05412","4":"0.6909696","5":"1.665603","6":"0.000000000","7":"0.0085018114","8":"91","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05412.png","10":"26","11":"AMEX60DD039603;AMEX60DD012164;AMEX60DD025304;AMEX60DD031139;AMEX60DD039599;AMEX60DD055791;AMEX60DD039606;AMEX60DD038318;AMEX60DD015201;AMEX60DD025721;AMEX60DD030862;AMEX60DD034544;AMEX60DD026889;AMEX60DD056342;AMEX60DD012123;AMEX60DD026892;AMEX60DD024402;AMEX60DD010084;AMEX60DDU001007100;AMEX60DD035296;AMEX60DD042070;AMEX60DD009940;AMEX60DD023720;AMEX60DD024110;AMEX60DD026906;AMEX60DD049332"},{"1":"map00513","2":"Various types of N-glycan biosynthesis","3":"https://www.genome.jp/dbget-bin/www_bget?map00513","4":"0.7070850","5":"1.653832","6":"0.000000000","7":"0.0109186570","8":"63","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00513.png","10":"20","11":"AMEX60DD004640;AMEX60DD004647;AMEX60DD004644;AMEX60DD041117;AMEX60DD004651;AMEX60DD004834;AMEX60DD041108;AMEX60DD004650;AMEX60DD008529;AMEX60DD006904;AMEX60DD042400;AMEX60DD014059;AMEX60DD034145;AMEX60DD034150;AMEX60DD049559;AMEX60DD042398;AMEX60DD015725;AMEX60DD003872;AMEX60DD050882;AMEX60DDU001011346"},{"1":"map00512","2":"Mucin type O-glycan biosynthesis","3":"https://www.genome.jp/dbget-bin/www_bget?map00512","4":"0.7286680","5":"1.633017","6":"0.000000000","7":"0.0150381504","8":"44","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00512.png","10":"16","11":"AMEX60DD013837;AMEX60DD012715;AMEX60DD031336;AMEX60DD031334;AMEX60DD031318;AMEX60DD031335;AMEX60DD029631;AMEX60DD055731;AMEX60DD030791;AMEX60DD055843;AMEX60DDU001007639;AMEX60DD013845;AMEX60DD004108;AMEX60DD012980;AMEX60DDU001002216;AMEX60DD045289"},{"1":"map05416","2":"Viral myocarditis","3":"https://www.genome.jp/dbget-bin/www_bget?map05416","4":"0.6553259","5":"1.604736","6":"0.000000000","7":"0.0215973436","8":"112","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05416.png","10":"36","11":"AMEX60DD021112;AMEX60DD012164;AMEX60DD009362;AMEX60DD009360;AMEX60DD021111;AMEX60DD025304;AMEX60DD009431;AMEX60DD009364;AMEX60DD015756;AMEX60DD029334;AMEX60DD009525;AMEX60DD025721;AMEX60DD034544;AMEX60DD009353;AMEX60DD014609;AMEX60DD009365;AMEX60DD009388;AMEX60DD016515;AMEX60DD009354;AMEX60DD026400;AMEX60DD021091;AMEX60DD028265;AMEX60DD009940;AMEX60DD009427;AMEX60DD049332;AMEX60DD009395;AMEX60DD028266;AMEX60DD047074;AMEX60DD028267;AMEX60DD053160;AMEX60DD009358;AMEX60DD009400;AMEX60DD009396;AMEX60DD048456;AMEX60DD030542;AMEX60DD030543"},{"1":"map05410","2":"Hypertrophic cardiomyopathy","3":"https://www.genome.jp/dbget-bin/www_bget?map05410","4":"0.6646096","5":"1.611760","6":"0.000000000","7":"0.0217893200","8":"95","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05410.png","10":"31","11":"AMEX60DD012164;AMEX60DD008613;AMEX60DD025304;AMEX60DD054502;AMEX60DD004521;AMEX60DD005313;AMEX60DD031139;AMEX60DD055791;AMEX60DD015201;AMEX60DD025721;AMEX60DD030862;AMEX60DD034544;AMEX60DD056342;AMEX60DD012123;AMEX60DD024402;AMEX60DDU001007100;AMEX60DD010588;AMEX60DD035296;AMEX60DD042070;AMEX60DD009940;AMEX60DD023720;AMEX60DDU001036307;AMEX60DD024110;AMEX60DD049332;AMEX60DD010587;AMEX60DD025328;AMEX60DD017438;AMEX60DD055675;AMEX60DD048456;AMEX60DD006925;AMEX60DD006923"},{"1":"map04530","2":"Tight junction","3":"https://www.genome.jp/dbget-bin/www_bget?map04530","4":"0.6389239","5":"1.605743","6":"0.000000000","7":"0.0227753805","8":"262","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map04530.png","10":"61","11":"AMEX60DD021112;AMEX60DD055382;AMEX60DD012164;AMEX60DD021111;AMEX60DD025304;AMEX60DD048585;AMEX60DD054502;AMEX60DD054265;AMEX60DD047355;AMEX60DD012944;AMEX60DD042464;AMEX60DD019404;AMEX60DD029334;AMEX60DD028403;AMEX60DD016873;AMEX60DD043649;AMEX60DD032116;AMEX60DD038631;AMEX60DD009525;AMEX60DD019402;AMEX60DD003189;AMEX60DD029929;AMEX60DD007292;AMEX60DD025721;AMEX60DD027428;AMEX60DD043143;AMEX60DD028031;AMEX60DD034544;AMEX60DD054267;AMEX60DD043142;AMEX60DD015214;AMEX60DD042465;AMEX60DD038976;AMEX60DD016962;AMEX60DD014337;AMEX60DD035181;AMEX60DDU001012094;AMEX60DD011843;AMEX60DD016515;AMEX60DD043146;AMEX60DD021091;AMEX60DD054271;AMEX60DD029357;AMEX60DD006620;AMEX60DD054262;AMEX60DDU001036307;AMEX60DD022047;AMEX60DDU001007247;AMEX60DD007874;AMEX60DD030882;AMEX60DD023658;AMEX60DD003190;AMEX60DD034849;AMEX60DD034850;AMEX60DD014940;AMEX60DD024266;AMEX60DD006925;AMEX60DD021574;AMEX60DD030542;AMEX60DD051325;AMEX60DD006923"},{"1":"map04961","2":"Endocrine and other factor-regulated calcium reabsorption","3":"https://www.genome.jp/dbget-bin/www_bget?map04961","4":"0.6630312","5":"1.582773","6":"0.001059322","7":"0.0280422652","8":"75","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map04961.png","10":"19","11":"AMEX60DD053460;AMEX60DD053461;AMEX60DD012899;AMEX60DD026384;AMEX60DD026388;AMEX60DDU001037010;AMEX60DD018138;AMEX60DD026390;AMEX60DD016809;AMEX60DD013795;AMEX60DD026385;AMEX60DD053503;AMEX60DD035741;AMEX60DD016805;AMEX60DD020537;AMEX60DD031148;AMEX60DD032746;AMEX60DD029429;AMEX60DD021663"},{"1":"map05140","2":"Leishmaniasis","3":"https://www.genome.jp/dbget-bin/www_bget?map05140","4":"0.6474167","5":"1.583324","6":"0.000000000","7":"0.0299778513","8":"105","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05140.png","10":"36","11":"AMEX60DD009362;AMEX60DD009360;AMEX60DD009431;AMEX60DD009364;AMEX60DD054519;AMEX60DD054520;AMEX60DD037755;AMEX60DD009353;AMEX60DD008808;AMEX60DD008986;AMEX60DD009365;AMEX60DD032075;AMEX60DD010469;AMEX60DD009388;AMEX60DD044244;AMEX60DD010588;AMEX60DD050615;AMEX60DD009354;AMEX60DD030024;AMEX60DD023202;AMEX60DD009427;AMEX60DD009395;AMEX60DD010587;AMEX60DD001287;AMEX60DD025328;AMEX60DD026167;AMEX60DD001084;AMEX60DD031148;AMEX60DD032076;AMEX60DD047292;AMEX60DD009358;AMEX60DD051083;AMEX60DD009400;AMEX60DD026163;AMEX60DD009396;AMEX60DD031847"},{"1":"map04978","2":"Mineral absorption","3":"https://www.genome.jp/dbget-bin/www_bget?map04978","4":"0.6638059","5":"1.547954","6":"0.001097695","7":"0.0371954251","8":"63","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map04978.png","10":"17","11":"AMEX60DD014989;AMEX60DD053460;AMEX60DD053461;AMEX60DD012899;AMEX60DD026384;AMEX60DD026388;AMEX60DDU001037010;AMEX60DD018138;AMEX60DD042385;AMEX60DD042374;AMEX60DD026390;AMEX60DD048093;AMEX60DD009033;AMEX60DD026385;AMEX60DD041025;AMEX60DD053503;AMEX60DD009014"},{"1":"map00830","2":"Retinol metabolism","3":"https://www.genome.jp/dbget-bin/www_bget?map00830","4":"0.6513521","5":"1.559028","6":"0.000000000","7":"0.0374953882","8":"78","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00830.png","10":"39","11":"AMEX60DD036247;AMEX60DD009506;AMEX60DD054954;AMEX60DD044208;AMEX60DD044196;AMEX60DD054953;AMEX60DD055500;AMEX60DD044195;AMEX60DD024496;AMEX60DD024495;AMEX60DD009933;AMEX60DD044203;AMEX60DD043999;AMEX60DD024364;AMEX60DD055711;AMEX60DD030654;AMEX60DDU001003190;AMEX60DD008898;AMEX60DD006986;AMEX60DD019294;AMEX60DD008899;AMEX60DD019729;AMEX60DD001380;AMEX60DD008900;AMEX60DD044207;AMEX60DD043987;AMEX60DD044795;AMEX60DD046276;AMEX60DD024492;AMEX60DD004101;AMEX60DD029708;AMEX60DD043997;AMEX60DD029693;AMEX60DD024499;AMEX60DD024485;AMEX60DD024552;AMEX60DD025756;AMEX60DD024493;AMEX60DD025757"},{"1":"map00350","2":"Tyrosine metabolism","3":"https://www.genome.jp/dbget-bin/www_bget?map00350","4":"0.6701848","5":"1.549364","6":"0.003351955","7":"0.0379405966","8":"52","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00350.png","10":"21","11":"AMEX60DD036247;AMEX60DD050176;AMEX60DD026518;AMEX60DD054954;AMEX60DD044208;AMEX60DD044196;AMEX60DD054953;AMEX60DD044195;AMEX60DD044203;AMEX60DD043999;AMEX60DD048606;AMEX60DD000630;AMEX60DD043361;AMEX60DD038234;AMEX60DD044207;AMEX60DD028229;AMEX60DD043997;AMEX60DD030183;AMEX60DD047599;AMEX60DD047598;AMEX60DDU001022505"},{"1":"map04261","2":"Adrenergic signaling in cardiomyocytes","3":"https://www.genome.jp/dbget-bin/www_bget?map04261","4":"0.6223717","5":"1.554550","6":"0.000000000","7":"0.0381129593","8":"180","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map04261.png","10":"38","11":"AMEX60DD021112;AMEX60DD012164;AMEX60DD008613;AMEX60DD021111;AMEX60DD025304;AMEX60DD054502;AMEX60DD004521;AMEX60DD012899;AMEX60DD023295;AMEX60DD018138;AMEX60DD031139;AMEX60DD001307;AMEX60DD048093;AMEX60DD013795;AMEX60DD009525;AMEX60DD009714;AMEX60DD034544;AMEX60DD030139;AMEX60DD021356;AMEX60DD012123;AMEX60DD008808;AMEX60DD053503;AMEX60DD047253;AMEX60DD024402;AMEX60DDU001007100;AMEX60DD015043;AMEX60DD009636;AMEX60DD016515;AMEX60DD009634;AMEX60DD021091;AMEX60DD002404;AMEX60DD035296;AMEX60DD029440;AMEX60DD023202;AMEX60DD035741;AMEX60DD023720;AMEX60DD024110;AMEX60DD021966"},{"1":"map00750","2":"Vitamin B6 metabolism","3":"https://www.genome.jp/dbget-bin/www_bget?map00750","4":"0.8517501","5":"1.551456","6":"0.004291845","7":"0.0383952775","8":"10","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00750.png","10":"4","11":"AMEX60DD036247;AMEX60DD054954;AMEX60DD054953;AMEX60DD029357"},{"1":"map05322","2":"Systemic lupus erythematosus","3":"https://www.genome.jp/dbget-bin/www_bget?map05322","4":"0.6278431","5":"1.560240","6":"0.000000000","7":"0.0393551594","8":"140","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05322.png","10":"74","11":"AMEX60DD009362;AMEX60DD009360;AMEX60DD009431;AMEX60DD009364;AMEX60DDU001011163;AMEX60DD008362;AMEX60DD009353;AMEX60DD048151;AMEX60DD017385;AMEX60DD009365;AMEX60DD032075;AMEX60DD010469;AMEX60DD022890;AMEX60DD009388;AMEX60DD010588;AMEX60DDU001037016;AMEX60DD010639;AMEX60DD050615;AMEX60DD009354;AMEX60DD026400;AMEX60DD017383;AMEX60DD010652;AMEX60DD046381;AMEX60DD010555;AMEX60DD046389;AMEX60DD017384;AMEX60DD009427;AMEX60DD010445;AMEX60DD010437;AMEX60DD046665;AMEX60DDU001032689;AMEX60DD021489;AMEX60DD009395;AMEX60DD010653;AMEX60DD010587;AMEX60DD010682;AMEX60DD026774;AMEX60DD047074;AMEX60DD041970;AMEX60DD026773;AMEX60DD032076;AMEX60DD009462;AMEX60DD009358;AMEX60DD026431;AMEX60DD009400;AMEX60DDU001038209;AMEX60DD010565;AMEX60DD024266;AMEX60DD009396;AMEX60DD010681;AMEX60DDU001018529;AMEX60DD031847;AMEX60DD026430;AMEX60DD026775;AMEX60DD038932;AMEX60DD008446;AMEX60DD010561;AMEX60DD014616;AMEX60DD010430;AMEX60DD010431;AMEX60DD032184;AMEX60DD010571;AMEX60DD010428;AMEX60DD031844;AMEX60DD041969;AMEX60DD010523;AMEX60DD031848;AMEX60DD010545;AMEX60DD031846;AMEX60DD041582;AMEX60DD013627;AMEX60DD046322;AMEX60DDU001031441;AMEX60DD009366"},{"1":"map00220","2":"Arginine biosynthesis","3":"https://www.genome.jp/dbget-bin/www_bget?map00220","4":"0.7137284","5":"1.534627","6":"0.012224939","7":"0.0418174653","8":"28","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00220.png","10":"14","11":"AMEX60DD018629;AMEX60DD054519;AMEX60DD054520;AMEX60DD011805;AMEX60DD029866;AMEX60DD040714;AMEX60DD050366;AMEX60DD026167;AMEX60DD001084;AMEX60DD050365;AMEX60DD026163;AMEX60DD054302;AMEX60DD026168;AMEX60DD040713"},{"1":"map05310","2":"Asthma","3":"https://www.genome.jp/dbget-bin/www_bget?map05310","4":"0.6892369","5":"1.535188","6":"0.008215962","7":"0.0431946872","8":"36","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05310.png","10":"19","11":"AMEX60DD009362;AMEX60DD009360;AMEX60DD009431;AMEX60DD009364;AMEX60DD029384;AMEX60DD009353;AMEX60DD009365;AMEX60DD009388;AMEX60DD010588;AMEX60DD009354;AMEX60DD026400;AMEX60DD009427;AMEX60DD009395;AMEX60DD010587;AMEX60DD009358;AMEX60DD009400;AMEX60DD009396;AMEX60DD026775;AMEX60DD008446"},{"1":"map04672","2":"Intestinal immune network for IgA production","3":"https://www.genome.jp/dbget-bin/www_bget?map04672","4":"0.6593457","5":"1.530226","6":"0.004390779","7":"0.0435146478","8":"56","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map04672.png","10":"24","11":"AMEX60DD009362;AMEX60DD009360;AMEX60DD009431;AMEX60DD009364;AMEX60DD042005;AMEX60DD041536;AMEX60DD009353;AMEX60DD009365;AMEX60DD009388;AMEX60DD031801;AMEX60DD009354;AMEX60DD026400;AMEX60DD009427;AMEX60DD009395;AMEX60DD013690;AMEX60DD047074;AMEX60DD026882;AMEX60DD009358;AMEX60DD009400;AMEX60DD055020;AMEX60DD009396;AMEX60DD026775;AMEX60DD008446;AMEX60DDU001000351"},{"1":"map00590","2":"Arachidonic acid metabolism","3":"https://www.genome.jp/dbget-bin/www_bget?map00590","4":"0.6340455","5":"1.535601","6":"0.001053741","7":"0.0450230337","8":"90","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map00590.png","10":"35","11":"AMEX60DD015066;AMEX60DD013313;AMEX60DD047232;AMEX60DD011931;AMEX60DD051734;AMEX60DD026437;AMEX60DDU001033114;AMEX60DD023020;AMEX60DD024495;AMEX60DD013438;AMEX60DD013437;AMEX60DD010596;AMEX60DD024364;AMEX60DD013324;AMEX60DD013300;AMEX60DD013130;AMEX60DD013309;AMEX60DD030654;AMEX60DD019428;AMEX60DDU001003190;AMEX60DD008898;AMEX60DD006986;AMEX60DD050938;AMEX60DD013307;AMEX60DD008899;AMEX60DD019729;AMEX60DD008900;AMEX60DD019427;AMEX60DD024492;AMEX60DD051083;AMEX60DD027558;AMEX60DD024499;AMEX60DD013439;AMEX60DD024485;AMEX60DD024552"},{"1":"map05146","2":"Amoebiasis","3":"https://www.genome.jp/dbget-bin/www_bget?map05146","4":"0.6184339","5":"1.521439","6":"0.000000000","7":"0.0485700260","8":"163","9":"./Project_Axolotl_GSEA_KEGG_GSEA/map05146.png","10":"53","11":"AMEX60DD004446;AMEX60DD009362;AMEX60DD009360;AMEX60DD004448;AMEX60DD009431;AMEX60DD009364;AMEX60DD018597;AMEX60DD054519;AMEX60DD054443;AMEX60DD054520;AMEX60DD038403;AMEX60DD013795;AMEX60DD008951;AMEX60DD004450;AMEX60DD039516;AMEX60DD009353;AMEX60DD037143;AMEX60DD009365;AMEX60DD010491;AMEX60DD011805;AMEX60DD037141;AMEX60DD009388;AMEX60DD044244;AMEX60DD010588;AMEX60DD009354;AMEX60DD030024;AMEX60DD002782;AMEX60DD035741;AMEX60DD004497;AMEX60DD009427;AMEX60DD026932;AMEX60DD023300;AMEX60DD009395;AMEX60DD010587;AMEX60DD001287;AMEX60DD047874;AMEX60DD025328;AMEX60DD010831;AMEX60DD047873;AMEX60DD038399;AMEX60DD026167;AMEX60DD001084;AMEX60DD031148;AMEX60DD032746;AMEX60DD053160;AMEX60DDU001040729;AMEX60DD009358;AMEX60DD009400;AMEX60DD026163;AMEX60DD029426;AMEX60DD024266;AMEX60DD009396;AMEX60DD002444"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Open the HTML report
<code>WebGestaltR_results/Project_Axolotl_GSEA_KEGG/Report_Axolotl_GSEA_KEGG.html</code>
from the files pane in a browser.</p>
<p>Expand ‘Job summary’ to read that “22 positive related categories and
no negative related categories” are significant in this analysis. This
is in contrast to the one negative category we observed when running
KEGG GSEA with <code>clusterProfiler</code>. We expect some differences
between these tools.</p>
<p>Compare the tablar results in this report to the treeplot we produced
under code chunk <code>treeplot CP KEGG GSEA</code>. There are a lot of
shared terms, and this is reassuring.</p>
</div>
</div>
<div id="save-versions-and-session-details" class="section level1">
<h1>7. Save versions and session details</h1>
<p>Both <code>clusterProfiler</code> and <code>WebGestaltR</code> query
the latest versions of the databases, with no function to list the
database version.</p>
<p>For this reason, we will save the analysis date to our rendered
notebook, so the external database version could be back-calculated from
the date if required:</p>
<pre class="r"><code>cat(&quot;Date of analysis:\n&quot;)</code></pre>
<pre><code>## Date of analysis:</code></pre>
<pre class="r"><code>print(Sys.Date())</code></pre>
<pre><code>## [1] &quot;2024-11-18&quot;</code></pre>
<p>R version and package versions:</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.4.2 (2024-10-31)
## Platform: x86_64-pc-linux-gnu
## Running under: Ubuntu 22.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0
## 
## locale:
##  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
##  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
##  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Etc/UTC
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] enrichplot_1.26.2      WebGestaltR_0.4.6      clusterProfiler_4.14.3
##  [4] lubridate_1.9.3        forcats_1.0.0          stringr_1.5.1         
##  [7] purrr_1.0.2            tidyr_1.3.1            tibble_3.2.1          
## [10] ggplot2_3.5.1          tidyverse_2.0.0        ontologyIndex_2.12    
## [13] dplyr_1.1.4            readr_2.1.5           
## 
## loaded via a namespace (and not attached):
##   [1] DBI_1.2.2               gson_0.1.0              rlang_1.1.3            
##   [4] magrittr_2.0.3          DOSE_4.0.0              compiler_4.4.2         
##   [7] RSQLite_2.3.7           systemfonts_1.0.6       png_0.1-8              
##  [10] vctrs_0.6.5             reshape2_1.4.4          pkgconfig_2.0.3        
##  [13] crayon_1.5.2            fastmap_1.1.1           XVector_0.46.0         
##  [16] labeling_0.4.3          utf8_1.2.4              rmarkdown_2.26         
##  [19] tzdb_0.4.0              UCSC.utils_1.2.0        bit_4.0.5              
##  [22] xfun_0.43               zlibbioc_1.52.0         cachem_1.0.8           
##  [25] aplot_0.2.3             GenomeInfoDb_1.42.0     jsonlite_1.8.8         
##  [28] blob_1.2.4              highr_0.10              BiocParallel_1.40.0    
##  [31] parallel_4.4.2          R6_2.5.1                bslib_0.7.0            
##  [34] stringi_1.8.4           RColorBrewer_1.1-3      jquerylib_0.1.4        
##  [37] GOSemSim_2.32.0         iterators_1.0.14        Rcpp_1.0.12            
##  [40] knitr_1.46              ggtangle_0.0.4          R.utils_2.12.3         
##  [43] IRanges_2.40.0          Matrix_1.7-1            splines_4.4.2          
##  [46] igraph_2.0.3            timechange_0.3.0        tidyselect_1.2.1       
##  [49] qvalue_2.38.0           rstudioapi_0.16.0       yaml_2.3.8             
##  [52] doParallel_1.0.17       codetools_0.2-19        curl_5.2.1             
##  [55] doRNG_1.8.6             lattice_0.22-5          plyr_1.8.9             
##  [58] treeio_1.30.0           Biobase_2.66.0          withr_3.0.0            
##  [61] KEGGREST_1.46.0         evaluate_0.23           gridGraphics_0.5-1     
##  [64] Biostrings_2.74.0       ggtree_3.14.0           pillar_1.9.0           
##  [67] rngtools_1.5.2          whisker_0.4.1           foreach_1.5.2          
##  [70] stats4_4.4.2            ggfun_0.1.7             generics_0.1.3         
##  [73] vroom_1.6.5             S4Vectors_0.44.0        hms_1.1.3              
##  [76] tidytree_0.4.6          munsell_0.5.1           scales_1.3.0           
##  [79] apcluster_1.4.13        glue_1.7.0              lazyeval_0.2.2         
##  [82] tools_4.4.2             ggnewscale_0.5.0        data.table_1.15.4      
##  [85] fgsea_1.32.0            fs_1.6.4                fastmatch_1.1-4        
##  [88] cowplot_1.1.3           grid_4.4.2              ape_5.8                
##  [91] AnnotationDbi_1.68.0    colorspace_2.1-0        nlme_3.1-165           
##  [94] GenomeInfoDbData_1.2.13 patchwork_1.3.0         cli_3.6.2              
##  [97] fansi_1.0.6             svglite_2.1.3           gtable_0.3.5           
## [100] R.methodsS3_1.8.2       yulab.utils_0.1.8       sass_0.4.9             
## [103] digest_0.6.35           BiocGenerics_0.52.0     ggrepel_0.9.6          
## [106] ggplotify_0.1.2         farver_2.1.2            memoise_2.0.1          
## [109] htmltools_0.5.8.1       R.oo_1.26.0             lifecycle_1.0.4        
## [112] httr_1.4.7              GO.db_3.20.0            bit64_4.0.5</code></pre>
<p>And RStudio version. Typically, we would simply run
<code>RStudio.Version()</code> to print the version details. However,
when we knit this document to HTML, the <code>RStudio.Version()</code>
function is not available and will cause an error. So to make sure our
version details are saved to our static record of the work, we will save
to a file, then print the file contents back into the notebook.</p>
<pre class="r"><code># Get RStudio version information
rstudio_info &lt;- RStudio.Version()

# Convert the version information to a string
rstudio_version_str &lt;- paste(
  &quot;RStudio Version Information:\n&quot;,
  &quot;Version: &quot;, rstudio_info$version, &quot;\n&quot;,
  &quot;Release Name: &quot;, rstudio_info$release_name, &quot;\n&quot;,
  &quot;Long Version: &quot;, rstudio_info$long_version, &quot;\n&quot;,
  &quot;Mode: &quot;, rstudio_info$mode, &quot;\n&quot;,
  &quot;Citation: &quot;, rstudio_info$citation,
  sep = &quot;&quot;
)

# Write the output to a text file
writeLines(rstudio_version_str, &quot;rstudio_version.txt&quot;)</code></pre>
<pre class="r"><code># Read the saved version information from the file
rstudio_version_text &lt;- readLines(&quot;rstudio_version.txt&quot;)

# Print the version information to the document
rstudio_version_text</code></pre>
<pre><code>## [1] &quot;RStudio Version Information:&quot;                                                                                                                                                                                                                                                                           
## [2] &quot;Version: 2023.6.1.524&quot;                                                                                                                                                                                                                                                                                  
## [3] &quot;Release Name: Mountain Hydrangea&quot;                                                                                                                                                                                                                                                                       
## [4] &quot;Long Version: 2023.06.1+524&quot;                                                                                                                                                                                                                                                                            
## [5] &quot;Mode: server&quot;                                                                                                                                                                                                                                                                                           
## [6] &quot;Citation: list(title = \&quot;RStudio: Integrated Development Environment for R\&quot;, author = list(list(given = \&quot;Posit team\&quot;, family = NULL, role = NULL, email = NULL, comment = NULL)), organization = \&quot;Posit Software, PBC\&quot;, address = \&quot;Boston, MA\&quot;, year = \&quot;2023\&quot;, url = \&quot;http://www.posit.co/\&quot;)&quot;</code></pre>
</div>
<div id="knit-workbook-to-html" class="section level1">
<h1>8. Knit workbook to HTML</h1>
<p>The last task is to knit the notebook. Our notebook is editable, and
can be changed. Deleting code deletes the output, so we could lose
valuable details. If we knit the notebook to HTML, we have a permanent
static copy of the work.</p>
<p>On the editor pane toolbar, under Preview, select Knit to HTML.</p>
<p>If you have already run Preview, you will see Knit instead of
Preview.</p>
<p>The HTML file will be saved in the same directory as the notebook,
and with the same filename, but the .Rmd prefix will be replaced by
.html. The knit HTML will typically open automatically once complete. If
you receive a popup blocker error, click cancel, and in the Files pane
of RStudio, single click the gprofiler.html file and select View in Web
Browser.</p>
<p>Note that the notebook will only successfully knit if there are no
errors in the code. You can ‘preview’ HTML with code errors.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
