---
title: "Enrichment with clusterProfiler"
output: html_notebook
---

```{r Load R packages, include=FALSE}
library(clusterProfiler)
library(enrichplot)
library(readr)
library(biomaRt)
library(org.Hs.eg.db)
library(dplyr)
library(ggupset)
```

Review the contents page of the [clusterProfiler user guide](https://bioconductor.org/packages/devel/bioc/manuals/clusterProfiler/man/clusterProfiler.pdf). 

`clusterProfiler` has a number of GSEA and ORA functions specifically designed to perform GSEA on key databases without having to manually import the database. 

For GSEA, these are:

- `gseGo`
  - Database: GO
  - Supported organisms: 
- `gseKEGG`
  - Database: KEGG
  - Supported organisms: all KEGG organisms (thousands) 
- `gseMKEGG`
  - Database: KEGG Modules
  - Supported organisms: 
- `gsePC`
  - Database: Pathway Commons
  - Supported organisms: 
- `gseWP`
  - Database: WikiPathways
  - Supported organisms: 
  
  
For ORA, the same 5 databases are supported, eg function `enrichGO` for ORA of GO, with the addition of `enrichDAVID`. 

Note that each of the in-built database functions has different supported organisms, and this is to do with the database itself. There are also limitations on which namespaces a database accepts, so you need to check the compatible namespace for the function you want to run, and perform gene ID conversion if required. `clusterProfiler` has the function `bitr` to do this; another popular choice is `biomaRt`. 


`clusterProfiler` also has generic functions to perform ORA (`enricher`) or GSEA (`gsea`) with custom gene sets. These enable you to perform ORA or GSEA with your own annotations, with a little work on your part in preparing the required database files. 




# 1. Load input data and extract ranked gene list

We will use the same Pezzini RNAseq dataset as earlier. First, confirm your working directory is the `Functional_enrichment_workshop_2024/day_2` folder. 

```{r setup, include=FALSE}
# Set the root directory for all notebook chunks
knitr::opts_knit$set(root.dir = "C:/Users/cwil2281/OneDrive - The University of Sydney (Staff)/Documents/PIPE-5119-enrichment-workshop/day_2")

# To set the working directory for the console, run this in the console itself:
setwd("C:/Users/cwil2281/OneDrive - The University of Sydney (Staff)/Documents/PIPE-5119-enrichment-workshop/day_2")


getwd()

```


Please ask for assistance if this step does not show the day_2 folder path! 

Once your working directory is set, load the dataset and extract the gene lists.

```{r load input data}

# Full dataset
data <- read_tsv("Pezzini_DE.txt", col_names = TRUE, show_col_types = FALSE)

# Ranked gene list vector for GSEA
ranked <- setNames(data$Log2FC[order(-data$Log2FC)], data$Gene.ID[order(-data$Log2FC)])

head(ranked)
tail(ranked)

```


# 2. Perform GSEA of GO terms with `gseGO` function 

Since we have covered ORA with `gprofiler`. 

We will run GSEA over the GO database. If we choose `ALL` for the `ont` (ontology) parameter,  we can run over MF, BP and CC in one run. However for simplicity, let's choose just `MF`.  

## Species

For this function, we need an [annotation database from Bioconductor](https://bioconductor.org/packages/3.20/data/annotation/). The package `org.Hs.eg.db` has been installed and loaded for you. 


## Namespace

We can use any gene namespace supported by the annotation package:

```{r org hs namespaces}
keytypes(org.Hs.eg.db)
```
We already have ENSEMBL gene IDs in our input data, so we don't need to do a gene ID conversion. 

For the rest of the parameters, we will use the defaults, except for `seed` that we should set to a specific value for reproducibility, and
add 10,000 permutations to make the analysis more robust and reliable by providing a better estimation of p-values for the enrichment scores, especially for cases with smaller gene sets or less significant differences.

## Run the enrichment

```{r gseGO}
gsea_GO <- gseGO( 
  ranked, 
  ont = "ALL", 
  OrgDb = "org.Hs.eg.db",
  keyType = "ENSEMBL", 
  exponent = 1, 
  minGSSize = 10, 
  maxGSSize = 500, 
  eps = 1e-10, 
  pvalueCutoff = 0.05, 
  pAdjustMethod = "BH", 
  verbose = TRUE, 
  seed = 1234, 
  by = "fgsea",
  nPermSimple = 10000
)

```


```{r gseGO MF }
gsea_GO_MF <- gseGO( 
  ranked, 
  ont = "MF", 
  OrgDb = "org.Hs.eg.db",
  keyType = "ENSEMBL", 
  exponent = 1, 
  minGSSize = 10, 
  maxGSSize = 500, 
  eps = 1e-10, 
  pvalueCutoff = 0.05, 
  pAdjustMethod = "BH", 
  verbose = TRUE, 
  seed = 1234, 
  by = "fgsea",
  nPermSimple = 10000
)

```


#3. Visualise  GSEA GO results 

Next we will visualise the GSEA results.

## As a table

Simple view: 89 enriched terms for GO: MF

```{r}
print(gsea_GO_MF)

```

Extract results to a TSV: this will print the significant enrichments, sorted by adjusted P value. 

```{r GSEA GO MF results table }

write.table(clsuterProfiler_gsea_GO_MF, file = "gsea_GO_MF_results.tsv", sep = "\t", quote = FALSE, row.names = FALSE)


```



## Dotplot 

this function is part of `lcusterProfiler` as well as `enrichPlot`. 

An easy to interpret visualisation, but  loses the details of the source `CC`, `MF`, `BP`. 


```{r GSEA GO dotplot }

# Create plot with dynamic title
p <- dotplot(
  gsea_GO,
  x = "geneRatio",
  color = "p.adjust",
  orderBy = "x",
  showCategory = 15,
  font.size = 8
  ) +
  ggtitle("GSEA GO")
    


print(p)

```

```{r GSEA GO MF dotplot }

# Create plot with dynamic title
p <- dotplot(
  gsea_GO_MF,
  x = "geneRatio",
  color = "p.adjust",
  orderBy = "x",
  showCategory = 15,
  font.size = 8
  ) +
  ggtitle("GSEA GO: Molecular function")
    


print(p)

```


## Upset plot

Shows the shared genes among the enriched terms (since a gene can belong to multiple terms or pathways).

For ORA, upsetplot will calculate the overlaps among different gene sets. 

For GSEA, it will plot the fold change distributions of different categories (e.g. unique to pathway, overlaps among different pathways).

`n` controls the number of terms plotted - her we have restricted to 10 for ease of viewing. Plot view is quite small within a notebook, but if wanted to plot more terms, you could print to an A4 size for enhanced clarity. 



```{r GSEA GO upset plot}

# n = the number of top enrichments to plot
p<- upsetplot(gsea_GO, n = 10)
print(p)

```


```{r GSEA GO MF upset plot}

# n = the number of top enrichments to plot
p<- upsetplot(gsea_GO_MF, n = 10)
print(p)

```


```{r GSEA GO upset plot print PDF }
# Make the plot, save to plot object 
p<- upsetplot(gsea_GO, n = 30)

# Open PDF in A4 landscape size in inches
pdf("GSEA_GO_upsetplot.pdf", width = 11.69, height = 8.27)  

# Print the plot to the device
print(p)

# Close the device (this saves the plot)
dev.off()
```


- Personally, i dislike upset plots, although they seem popular... 


# Enrichment map

Encrichment map with `emapplot` function from `clsuterProfiler`/`enrichPlot`: organizes enriched terms into a network with edges connecting overlapping gene sets. Overlapping gene sets cluster together.


```{r GSEA GO emapplot }
# calculate pairwise similarities between the enriched terms
gsea_GO <- pairwise_termsim(gsea_GO)

# plot
p<-emapplot(gsea_GO, showCategory = 15)
print(p)

```

```{r GSEA GO MF emapplot }
# calculate pairwise similarities between the enriched terms
gsea_GO_MF <- pairwise_termsim(gsea_GO_MF)

# plot
p<-emapplot(gsea_GO_MF, showCategory = 15)
print(p)

```



# cnet plot

From here: https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html 

"Both the barplot() and dotplot() only displayed most significant or selected enriched terms, while users may want to know which genes are involved in these significant terms. In order to consider the potentially biological complexities in which a gene may belong to multiple annotation categories and provide information of numeric changes if available, we developed the cnetplot() function to extract the complex association. The cnetplot() depicts the linkages of genes and biological concepts (e.g. GO terms or KEGG pathways) as a network. GSEA result is also supported with only core enriched genes displayed."

cnetplot: https://rdrr.io/bioc/enrichplot/man/cnetplot.html


```{r GSEA GO cnetplot}
# cex_label_gene to reduce the font size for gene ID
# colorEdge adds the coloured legend and network bars 

cnetplot(gsea_GO, showCategory = 15, cex_label_gene = 0.5, cex_label_category = 0.8, colorEdge = TRUE)
```


```{r GSEA GO MF cnetplot}

cnetplot( gsea_GO_MF, 
  showCategory = 15, 
  node_label = "category", 
)
```


- These look hideous and nothing like the last time I ran them on different dataset... 


## circular cnetplot 

 = no longer available in latest pacakge! damn they were so nice 

## treeplot

Functional grouping tree diagram (enrichPlot) 

```{r treeplot gSEA GO MF}
p<- treeplot(gsea_GO_MF, showCategory = 15, color = "p.adjust")

print(p)
```

## heatplot

```{r heatplot GSEA GO MF }
heatplot(gsea_GO_MF, showCategory = 15)
```
- Absolutely terrible, just no! 


# ridgeplot

This function is specific for GSEA (does not apply to ORA, since it plots NES) 

The height of the ridge in the ridge plot represents the density of genes from the gene set at different points in the ranked list.

The area under the curve represents the distribution of these genes across the ranked list. 

Higher peaks and more concentrated areas under the curve at the top of the list indicate stronger enrichment.


```{r ridgeplot  gsea GO MF, fig.height=10, fig.width=7}
ridgeplot(gsea_GO_MF, 
  showCategory = 30, 
  fill = "p.adjust", 
  core_enrichment = TRUE, 
  label_format = 30, 
  orderBy = "NES", 
  decreasing = FALSE )
```

## gseaplot

Another GSEA-specific plot. This will provide the GSEA plot for one term.

```{r gseaplot GSEA GO MF }

gene_set_id <- "GO:0008094"

p<- gseaplot( gsea_GO_MF,
  gene_set_id, 
  by = "all", 
  title = paste("GSEA result:", gene_set_id),
  color = "black", 
  color.line = "green", 
  color.vline = "#FA5860"
)

print(p)
```





#4. Perform GSEA on KEGG database with `gseKEGG` function 

Now we will run GSEA over another in-built database. Navigate to the page of the `clsuterProfiler` user guide where the `gseKEGG` function is described. 

## Species

Note that the default organism is human (`hsa`). 

If you were working with a species other than human, you first need to obtain your organism code. You can derive this from [KEGG Organisms](https://www.genome.jp/kegg/tables/br08606.html) or using the `clusterProfiler` function `search_kegg_organism`, for example `search_kegg_organism("horse", by = "common_name")`. 

Unlike `gseGO` that requires a Bioconductor annotation package, the `gseKEGG` function supports many more organisms. 

## Namespace

Required `key type` is one of `kegg`, `ncbi-geneid`, `ncib-proteinid` and `uniprot`. Our gene IDs are ENSEMBL. 

We can convert with `clusterProfiler` function `bitr`. Converting to ENTREZ will be compatible with `kegg`. 


```{r convert gene IDs}
# Convert the Ensembl IDs to Entrez IDs, dropping NAs
converted_ids <- bitr(names(ranked), 
                      fromType = "ENSEMBL", 
                      toType = "ENTREZID", 
                      OrgDb = org.Hs.eg.db, 
                      drop = TRUE)

# Handle any `1:many mapping` (ENSEMBL ID matches >1 ENTREZID) by keeping only the first Entrez ID for each Ensembl ID
converted_ids <- converted_ids[!duplicated(converted_ids$ENSEMBL), ]

# Keep only the rows with valid Entrez IDs
converted_ids <- converted_ids[!is.na(converted_ids$ENTREZID), ]

# Filter the ranked gene list to include only the successfully mapped Ensembl IDs
ranked_entrez <- ranked[names(ranked) %in% converted_ids$ENSEMBL]

# Replace the Ensembl IDs with the corresponding Entrez IDs in the ranked vector
names(ranked_entrez) <- converted_ids$ENTREZID[match(names(ranked_entrez), converted_ids$ENSEMBL)]

# Remove duplicates (keep only first occurrence of Entrez IDs)
## Note: this is not ideal and for a real experiment you should carefully review duplicates and choose which and how to retain based on your biological context 
ranked_entrez <- ranked_entrez[!duplicated(names(ranked_entrez))]

cat("\n")

# Check: 
head(ranked_entrez)
tail(ranked_entrez)

```

## Run the enrichment

Now run GSEA over the KEGG database with `gseKEGG` function: 


```{r gseKEGG}
gsea_kegg <- gseKEGG( 
  ranked_entrez, 
  organism = "hsa", 
  keyType = "kegg", 
  exponent = 1, 
  minGSSize = 10, 
  maxGSSize = 500, 
  eps = 1e-10, 
  pvalueCutoff = 0.05, 
  pAdjustMethod = "BH",
  verbose = TRUE, 
  use_internal_data = FALSE, 
  seed = FALSE, 
  by = "fgsea",
  nPermSimple = 10000
)


```


#5. Visualise  GSEA KEGG results 

First let's extract the tabular results: 


## As a table

40 enriched terms 

```{r}
print(gsea_kegg)

```

Extract results to a TSV: this will print the significant enrichments, sorted by adjusted P value. 

```{r GSEA GO MF results table }

write.table(gsea_kegg, file = "gsea_KEGG_results.tsv", sep = "\t", quote = FALSE, row.names = FALSE)


```

## Plotting - challenge!


- select your favourite plot from the plots generated for the `gseGO` MF results.
- create a new code chunk below and give it a unique chunk label
- paste the code for the plot into the code chunk
- edit the code to run on the `gseKEGG` output object
- run the code to p[roduce your plot! 
- What plot did you choose and why? Did ti work ok?



## Dotplot 



```{r GSEA KEGG dotplot }

# Create plot with dynamic title
p <- dotplot(
  gsea_kegg,
  x = "geneRatio",
  color = "p.adjust",
  orderBy = "x",
  showCategory = 15,
  font.size = 8
  ) +
  ggtitle("GSEA KEGG")
    

print(p)

```



## Upset plot

Shows the shared genes among the enriched terms (since a gene can belong to multiple terms or pathways).

For ORA, upsetplot will calculate the overlaps among different gene sets. 

For GSEA, it will plot the fold change distributions of different categories (e.g. unique to pathway, overlaps among different pathways).

`n` controls the number of terms plotted - her we have restricted to 10 for ease of viewing. Plot view is quite small within a notebook, but if wanted to plot more terms, you could print to an A4 size for enhanced clarity. 



```{r GSEA KEGG upset plot}

# n = the number of top enrichments to plot
p<- upsetplot(gsea_kegg, n = 10)
print(p)

```



# Enrichment map

Encrichment map with `emapplot` function from `clsuterProfiler`/`enrichPlot`: organizes enriched terms into a network with edges connecting overlapping gene sets. Overlapping gene sets cluster together.


```{r GSEA KEGG emapplot }
# calculate pairwise similarities between the enriched terms
gsea_kegg <- pairwise_termsim(gsea_kegg)

# plot
p<-emapplot(gsea_kegg, showCategory = 15)
print(p)

```



# cnet plot

From here: https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html 

"Both the barplot() and dotplot() only displayed most significant or selected enriched terms, while users may want to know which genes are involved in these significant terms. In order to consider the potentially biological complexities in which a gene may belong to multiple annotation categories and provide information of numeric changes if available, we developed the cnetplot() function to extract the complex association. The cnetplot() depicts the linkages of genes and biological concepts (e.g. GO terms or KEGG pathways) as a network. GSEA result is also supported with only core enriched genes displayed."

cnetplot: https://rdrr.io/bioc/enrichplot/man/cnetplot.html


```{r GSEA KEGG cnetplot}
# cex_label_gene to reduce the font size for gene ID
# colorEdge adds the coloured legend and network bars 

cnetplot(gsea_kegg, showCategory = 15, cex_label_gene = 0.5, cex_label_category = 0.8, colorEdge = TRUE)
```


- These look hideous and nothing like the last time I ran them on different dataset... 



## treeplot

Functional grouping tree diagram (enrichPlot) 

```{r treeplot GSEA KEGG}
p<- treeplot(gsea_kegg, showCategory = 15, color = "p.adjust")

print(p)
```

## heatplot

```{r heatplot GSEA KEGG }
heatplot(gsea_kegg, showCategory = 15)
```
- Absolutely terrible, just no! 


# ridgeplot

This function is specific for GSEA (does not apply to ORA, since it plots NES) 

The height of the ridge in the ridge plot represents the density of genes from the gene set at different points in the ranked list.

The area under the curve represents the distribution of these genes across the ranked list. 

Higher peaks and more concentrated areas under the curve at the top of the list indicate stronger enrichment.


```{r ridgeplot  gsea KEGG, fig.height=10, fig.width=7}
ridgeplot(gsea_kegg, 
  showCategory = 30, 
  fill = "p.adjust", 
  core_enrichment = TRUE, 
  label_format = 30, 
  orderBy = "NES", 
  decreasing = FALSE )
```

## gseaplot

Another GSEA-specific plot. This will provide the GSEA plot for one term.

```{r gseaplot GSEA GO MF }

gene_set_id <- "hsa04110"

p<- gseaplot( gsea_kegg,
  gene_set_id, 
  by = "all", 
  title = paste("GSEA result:", gene_set_id),
  color = "black", 
  color.line = "green", 
  color.vline = "#FA5860"
)

print(p)
```



# 6. Additional database support with `enricher` and `GSEA` functions

One of the great features of `clusterProfiler` is the ability to run ORA or GSEA over any gene set or database that isn't natively supported with the `gse<DB>` or `enrich<DB>` functions. This is through the use of the generic ORA and GSEA fnctions `enricher` and `GSEA`. The database and annotation details are provided to the functions through the `TERM2GENE` and `TERM2NAME` arguments. These are custom files that map your gene IDs to database terms. 

This enables extensive database support, as well as the option to analyse novel genomes. 

The `TERM2GENE` file is a 2-column list of gene IDs against database/gene set term IDs. There is one gene per line, and a gene can be on many lines ie be mapped to more than one term.

The `TERM2NAME` file will contain the ID of the term from the database, alongside its description, eg  `GO:0003676	nucleic acid binding`. 

So the `TERM2GENE` file is species-specific, while the `TERM2NAME` is generic, but subset to only include the terms that are present in the `TERM2GENE` file. 


 *** MAYBE THIS WILL BE INCLUDED IN NOTEBOOK 4 FOR NOVEL SPECIES - UNDECIDED WHETHER TO DO CLUSTERPROFILER, WEBGESTALTR, OR BOTH ***
 *** WEBGESTALTR IS AMAZING BUT I HAVE NOT YET TRIED ITS NON MODEL DB FILES ***
 *** IF IT WORKS, I WILL PROBABLY USE IT OVER CP. MAYBE CAN INCLUDE EXAMPEL CODE FOR BOTH SOMEWHERE *** 







