---
title: "ORA with gprofiler2"
output:
  html_document:
    df_print: paged
---

```{r Load R packages, include=FALSE}
library(readr)
library(dplyr)
library(gprofiler2)
library(ggplot2)
library(tidyr)

```

```{r check notebook workdir}
getwd()
```

```{r working directory, include=FALSE}

# Check default working directory: 
#getwd()

# Set the working directory for notebook
#knitr::opts_knit$set(root.dir = "workshop")
#getwd()

# Set working directory for console (copy paste into console)
#setwd("workshop")
#getwd()
```


# 1. Load input data

Raw data from [Pezzini et al 2017](https://link.springer.com/article/10.1007/s10571-016-0403-y) was subjected to differential gene expression analysis with [Degust](https://degust.erc.monash.edu/) and the results file saved to `Pezzini_DE.txt`.

The input data file is within the current working directory so we do not need to specify its directory path.  


```{r load input data}
data <- read_tsv("Pezzini_DE.txt", col_names = TRUE, show_col_types = FALSE)
head(data)
```

The dataframe shows genes with fold change and FDR values, along with some normalised counts values for the 6 samples (2 groups with 3 replicates each). 

Look on the environment pane of RStudio, and you can see a description '14420 obs. of 10 variables' - this shows your dataframe consists of 10 columns and 14,420 genes. 

# 2. Get ORA gene list

Now we need to filter for differentially expressed genes (DEGs), and we will apply the thresholds adjusted P values/FDR < 0.01, and log2fold change of 2. 

```{r get ORA gene list}
degs <- data %>%
  filter(FDR < 0.01 & abs(Log2FC) > 2) %>%
  pull(Gene.ID)
cat("Number of genes passing FDR and fold change filter:", length(degs), "\n")

# Save the DEG gene list to disk: 
write.table(degs, "Pezzini_DEGs.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)

```
We have 792 genes passing our filters. 

# 3. Get background gene list

Recall from the webinar and day 1 of the workshop that an experimental background gene list is crucial to avoiding false positives and minimising tissue bias with ORA. 

The analysis in Degust has already removed lowly expressed genes, so we can simply extract all genes from this data matrix as our background gene list and save it as our 'background' object, as well as save to disk so that we can include it within the supplementary materials of any resultant publications for reproducibility. 

```{r get background }
background <- data$Gene.ID
cat("Number of background genes:", length(background), "\n")

# Save the background gene list to disk:
write.table(background, "Pezzini_background_genes.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

# 4. Run ORA with `gost` function

Before running the below code chunk, review the parameters. Observe the similarities to what was available on the g:Profiler web interface, for example organism, the correction method (g:Profiler's custom `g_scs` method), and domain scope (background genes). 

Every R package has a user guide that outlines available parameters. Open the   [gprofiler2](https://cran.r-project.org/web/packages/gprofiler2/gprofiler2.pdf) user guide and navigate to page detailing the `gost` function (currently page 6). 

Note: to open this link for the RStudio training VM, highlightthe link, right click, then select "Go to https://cran..."

Under the `Usage` subheading, all available  parameters are listed, along with their default settings.  You can over-ride these parameters by specifying your custom values in your R command. 

Run the below code which explicitly includes all available `gost` parameters: 

```{r run gost}
ora <- gost( 
    degs, 
    organism = "hsapiens", 
    ordered_query = FALSE, 
    multi_query = FALSE, 
    significant = TRUE, 
    exclude_iea = FALSE, 
    measure_underrepresentation = FALSE, 
    evcodes = FALSE, 
    user_threshold = 0.05, 
    correction_method = "g_SCS", 
    domain_scope = "custom_annotated", 
    custom_bg = background, 
    numeric_ns = "", 
    sources = NULL, 
    as_short_link = FALSE, 
    highlight = FALSE
)
```


An error-free `gost` run should produce no console output.  Our results are saved in the R object `ora`.

Since we are using many of the default parameters, we could shorten this code to what is shown below. The results would be identical, however not as transparent as far as easily identifying what parameters were applied to a run. 

Tip: you can run the command `formals(gost)` to print out all parameters and their default values. This is a generic R function so can be used on other tools outside of `gprofiler2`. 


```{r abbreviated gost code }

#ora <- gost( 
#    degs, 
#    correction_method = "g_SCS", 
#    domain_scope = "custom_annotated",
#    custom_bg = background,
#)

```

```{r print defaults}
formals(gost)
```


View the top-most significant enrichments with the R `head` command. Only significant enrichments passing your specified threshold (adjusted P value < 0.05) are included in the results object.


```{r head ora}
head(ora$result)
```

Let's give our query a name:

```{r ora name the query list }
ora$result$query <- 'DEGs_Padj0.05_FC2'
head(ora$result)
```

Use the small black arrow near the column names to view columns wider than the page width.  The `head` view only shows the top 6 enrichments, which are all GO Biological Process. 

We can obtain a list of all databases with significant enrichments from our query list:


```{r ora dbs}
unique(ora$result$source)
```


Same as the web tool, we have enrichment results for GP (BP, CC, MF), HP (human phenotype), HPA (human protein atlas), KEGG, MiRNA, Reactome, Transcription Factors, and WikiPathways. 



# 5. Save the results to a file

Let's save the results file to disk. First, we will re-order the columns so the output more closely matches the tables that are downloaded from the web version of the tool.


```{r write ora table}

ora_reordered <- ora$result[, c("source", "term_name", "term_id", "p_value", "term_size", "query_size", "intersection_size", "effective_domain_size")]

head(ora_reordered)

write.csv(ora_reordered, "gprofiler_ORA_results.csv", row.names = FALSE)


```


# 6. Visualise the results 

## Manhattan plots 

The `gostplot` function creates a Manhattan plot similar to the one shown on the web tool. By setting `interactive=TRUE` we can hover over the data points to see enriched term details.

The parameter capped = TRUE is an indicator whether the -log10(p-values) would be capped at 16 if bigger than 16. This fixes the scale of y-axis to keep Manhattan plots from different queries comparable and is also intuitive as, statistically, p-values smaller than that can all be summarised as highly significant.


```{r gostplot}

gostplot(ora, 
  capped = TRUE, 
  interactive = TRUE, 
  pal = c(`GO:MF` = "#dc3912", 
    `GO:BP` = "#ff9900", 
    `GO:CC` = "#109618", 
    KEGG = "#dd4477", 
    REAC = "#3366cc", 
    WP = "#0099c6", 
    TF = "#5574a6", 
    MIRNA = "#22aa99", 
    HPA = "#6633cc", 
    CORUM = "#66aa00", 
    HP = "#990099") 
)

```


There are a lot of significant enrichments for GO biological processes. Many of these are probably terms containing a large number of genes, so not particularly informative. Other  R tools have default settings limiting the minimum and maximum number of genes in a geneset to be included in the analysis. Since there is no direct parameter to restrict term size to `gostplot`, we can filter the ORA results before plotting. Let's apply a maximum gene set size of 500, and a minimum gene set size of 10, which are the default setting used by clusterProfiler. 
 

```{r gostplot - filtered term size}
# Filter the results for GO:BP terms with term_size <= 500
ora_filter_termsize <- ora
ora_filter_termsize$result <- ora$result %>% filter(term_size <= 500) %>% filter(term_size >= 10)

# Plot with gostplot using the filtered results
gostplot(ora_filter_termsize, 
  capped = TRUE, 
  interactive = TRUE, 
  pal = c(
    `GO:MF` = "#dc3912", 
    `GO:BP` = "#ff9900", 
    `GO:CC` = "#109618", 
    KEGG = "#dd4477", 
    REAC = "#3366cc", 
    WP = "#0099c6", 
    TF = "#5574a6", 
    MIRNA = "#22aa99", 
    HPA = "#6633cc", 
    CORUM = "#66aa00", 
    HP = "#990099"
  )
)

```

This has cleaned up 'Biological Process' a little bit, enabling signals of more specific terms to be highlighted. 

gprofiler2 includes a function for creating a publication-ready image that can optionally highlight specific terms. We need to first produce a plot with `interactice = FALSE`, save it to an object, and then provide that plot object to the `publish_gostplot` function. 


```{r gostplot non-interactive}

# Plot with gostplot using the filtered results
plot <- gostplot(ora_filter_termsize, 
  capped = TRUE, 
  interactive = FALSE, 
  pal = c(
    `GO:MF` = "#dc3912", 
    `GO:BP` = "#ff9900", 
    `GO:CC` = "#109618", 
    KEGG = "#dd4477", 
    REAC = "#3366cc", 
    WP = "#0099c6", 
    TF = "#5574a6", 
    MIRNA = "#22aa99", 
    HPA = "#6633cc", 
    CORUM = "#66aa00", 
    HP = "#990099"
  )
)
```

The `publish_gostplot` parameter `highlight_terms` enables you to highlight specific terms on the plot, with a table showing enrichment details below for those highlighted terms. 

Let's highlight some selected terms manually. You need to provide the term ID not term name. 

```{r publish gostplot - collagen }
#Term IDs for 'Collagen degradation' and 'Collagen formation'
highlight <- c("REAC:R-HSA-1442490", "REAC:R-HSA-1474290")

filename <- "gostplot_filter_termsize_collagen.png"

publish_gostplot(plot, 
  highlight_terms = highlight, 
  filename, 
  width = 10, 
  height = 10 )

```


You can use R `grepl` function to search for terms with names matching some keyword. Let's highlight all terms related to receptors. The code chunk applies an increased figure height, to ensure we can see the whole plot within the notebook. 

```{r publish gostplot - receptors, fig.height=8}

# Filter terms containing "receptor" keyword and create a list of those term IDs
highlight <- ora$result %>% 
  filter(grepl("receptor", term_name, ignore.case = TRUE)) %>% 
  pull(term_id) 

filename <- "gostplot_filter_termsize_receptors.png"

publish_gostplot(plot, 
  highlight_terms = highlight, 
  filename, 
  width = 10, 
  height = 10 )

```



## Dotplots

One of the advantages of working in R is flexibility with visualisations. While the interactive Manhattan plots and `publish_gostplot` options are nice, it can also be useful to visualise P values aginst all term descriptions. 

One way to do this is with a dotplot. We can loop through all databases that `gost` has enriched against, and use the R package `ggplot2` to make a dotplot, first filtering the  results by database and(if desired)  term size:


```{r ora dotplots}
# List of databases
dbs <- unique(ora$result$source)

# Loop through each database and create a separate plo for each
for (db in dbs) {
  #db_results <- ora$result 
  db_results <- ora$result %>% filter(source == db, term_size >= 10, term_size <= 500)
  
  # Create a dot plot for the current database
  p <- ggplot(db_results, aes(x = reorder(term_name, -p_value), y = -log10(p_value))) +
    geom_point(aes(size = term_size, color = significant)) +
    labs(title = paste(db), 
         x = "Term", 
         y = "-log10(p-value)") +
    theme_minimal() +
    coord_flip()  # Flips the coordinates for better visibility
  
  # Print the plot
  print(p)
}
```

 
For plots with a lot of enriched terms, such as GO Biological Process, the display within the notebook is less than ideal. Saving the plot to an image file enables better resolution: 


```{r save go bp plot}

# Filter for the GO:BP database
go_bp_results <- ora$result %>% filter(source == "GO:BP",term_size >= 10, term_size <= 500)

# Create the plot for GO:BP
p_go_bp <- ggplot(go_bp_results, aes(x = reorder(term_name, -p_value), y = -log10(p_value))) +
  geom_point(aes(size = term_size, color = significant)) +
  labs(title = "GO:BP", 
       x = "Term", 
       y = "-log10(p-value)") +
  theme_minimal() +
  coord_flip()  # Flips the coordinates for better visibility


# Open a PDF device to save the plot as a full size A4: 
pdf("gprofiler_GO_BP_dotplot.pdf", width = 8.27, height = 11.69)  # A4 portrait size in inches

# Print the plot to the device
print(p_go_bp)

# Close the device (this saves the plot)
dev.off()

```
Open this plot by clicking it from the 'Files' pane of RStudio. Notice how the term names are now readable :-) 



# 7. Run a gost multi-query for up-regulated and down-regulated genes, and compare the results to those generated from the full DEG gene list

By providing more than one gene list and setting `multi_query = TRUE`, results from all of the gene lists are grouped by term IDs for easier comparison. This can be handy when you have multiple comparisons within an experiment, or when you want to investigate enrichments within the up and down regulated genes separately.  


First, we need to re-extract our gene list so we have separate DEGs for up-regulated and down-regulated genes. 

```{r get ORA gene lists up and down }
up_degs <- data %>%
  filter(FDR < 0.01 & Log2FC >= 2) %>%
  pull(Gene.ID)
cat("Number of upregulated DEGs:", length(up_degs), "\n")

# Save the DEG gene list to disk: 
write.table(up_degs, "up_DEGs.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)

down_degs <- data %>%
  filter(FDR < 0.01 & Log2FC <= -2) %>%
  pull(Gene.ID)
cat("Number of downregulated DEGs:", length(down_degs), "\n")

# Save the DEG gene list to disk: 
write.table(down_degs, "down_DEGs.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)

```

Now run `gost` as multi-query. The changes required for multi-query are providing a list of gene list objects to the `query` parameter instead of a single gene list object, and setting `multi_query = TRUE`, which is FALSE by default. By including all genes as well, we can efficintly compare up vs down vs no separation. 

```{r run gost multi-query }

ora_multi <- gost( 
    query = list("upregulated" = up_degs, "downregulated" = down_degs, "all_DEGs" = degs), 
    organism = "hsapiens", 
    ordered_query = FALSE, 
    multi_query = TRUE, 
    significant = TRUE, 
    exclude_iea = FALSE, 
    measure_underrepresentation = FALSE, 
    evcodes = FALSE, 
    user_threshold = 0.05, 
    correction_method = "g_SCS", 
    domain_scope = "custom_annotated", 
    custom_bg = background, 
    numeric_ns = "", 
    sources = NULL, 
    as_short_link = FALSE, 
    highlight = FALSE 
)
```

Now create a multi-query interactive Manhattan plot with `gostplot`:

```{r gostplot multi}
gostplot(ora_multi, capped = TRUE, interactive = TRUE)

```


Unfortunately, notebook view squashes the top plot over the bottom one, and adjusting figure height or plot layout options doesn't seem to help. Plotting as non-interactive or plotting from the console to the plots pane both produce a correct looking plot. 


```{r gostplot multi non-interactive}
p <- gostplot(ora_multi, capped = TRUE, interactive = FALSE)
filename <- "gprofiler_ORA_multiquery.png"

publish_gostplot(p, 
  highlight_terms = NULL, 
  filename, 
  width = 10, 
  height = 10 )
```


To access the tabular results separately, they need to be split, as the default results view 

***NOT WORKING***



```{r}

NR <- nrow(ora$result)
NS <- ifelse(NR < 10, NR, 10)
termSource <- sample(nrow(ora$result), NS)
VEC <- ora$result[termSource, "term_id"]

publish_gosttable(ora,
  highlight_terms = VEC,
  use_colors = TRUE,
  show_columns = c("source", "term_name", "term_size", "intersection_size"),
  filename = NULL)


```
Head ora multi in notebook: 

```{r}
head(ora_multi$result)
```


```{r}

NR <- nrow(ora_multi$result)
NS <- ifelse(NR < 10, NR, 10)
termSource <- sample(nrow(ora_multi$result), NS)
VEC <- ora_multi$result[termSource, "term_id"]

filename <- "publish-gosttable-multi.pdf"

publish_gosttable(ora,
  highlight_terms = VEC,
  use_colors = TRUE,
  show_columns = c("source", "term_name", "term_size", "intersection_size", "p_values", "significant"),
  filename)


```





***END NOT WORKING***

# 8. Compare gprofiler2 R results to the g:Profiler web results

In day 1 of the workshop, you ran ORA with g:Profiler web tool and saved the results to a CSV. Let's compare the results to those we have generated in R. Do we expect the results to be identical or differ slightly? 

The input file here is one that we have created, but should match yours as long as you used the same P filters and gprofiler parameters. 


```{r import g:profiler web results}
web <- read.csv("gProfiler_hsapiens_07-11-2024_11-27-09__intersections.csv")
```


Check the numbers: are there any terms significant frm one tool but not the other? 

```{r}
# Extract significant term names
web_terms <- web$term_name
ora_terms <- ora$result$term_name

paste0("Number of significant terms from web: ", length(web_terms))
paste0("Number of significant terms from R: ", length(ora_terms))

# Find command and unique terms
common_terms <- intersect(web_terms, ora_terms)
if (length(common_terms) == length(web_terms) && length(common_terms) == length(ora_terms)) {
  # If the lengths match, all terms are shared
  print("All terms are shared")
} else {
  # If there are differences, report the number of terms
  unique_web <- setdiff(web_terms, ora_terms)
  unique_ora <- setdiff(ora_terms, web_terms)
  
  print(paste("Number of terms unique to web:", length(unique_web)))
  print(paste("Number of terms unique to gprofiler2 (R):", length(unique_ora)))
}

```

That's a good start! Do the P values differ? Let's look closely at GO 'Molecular Function'. 

```{r fig.width=10, fig.height=8}

# Filter for GO:MF terms
go_mf_web <- web %>% filter(source == "GO:MF")
go_mf_r <- ora$result %>% filter(source == "GO:MF")

# Extract term names and p values
comparison_data_go_mf <- data.frame(
  term_name = go_mf_web$term_name,
  p_value_web = go_mf_web$adjusted_p_value,
  p_value_r = go_mf_r$p_value
)


# Reshape the data to long format
comparison_data_long <- comparison_data_go_mf %>%
  pivot_longer(cols = starts_with("p_value"), 
               names_to = "source", 
               values_to = "p_value")

# Create the bar plot with -log10 transformed p-values
print(ggplot(comparison_data_long, aes(x = term_name, y = -log10(p_value), fill = source)) +
  geom_bar(stat = "identity", position = "dodge", width = 0.7) +  # Side-by-side bars
  labs(title = "Adjusted P value comparison for GO:MF enrichments",
       x = "Term Name",
       y = "-log10(P-value)") +
  scale_fill_manual(values = c("p_value_web" = "#ff9900", "p_value_r" = "#3366cc")) +  # Custom colors
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) )  
```

Great! We know we applied the same parameters, and used the same input gene lists. The identical results must mean that gprofiler2 is using the same database version as g:Profiler web. 

# Saving versions and session details 

Let's check: yesterday when you ran ORA on the web, hopefully you saved your 'query parameters' as well as your results. 

From my run, I can see version as 'e111_eg58_p18_f463989d'. 

Let's report the g:Profiler database version used in our analysis:

```{r gprofiler versions}

paste0("g:Profiler database version: ", ora$meta$version)
paste0("gprofiler2 package version: ", packageVersion("gprofiler2"))

```

We can also capture the version of R and other session details including all loaded packages and versions with the `sessionInfo()` function: 

```{r info }
sessionInfo()
```

And finally, the RStudio version. Typically, we would simply run `RStudio.Version()` to print the version details. However, when we knit this document to HTML, the `RStudio.Version()` function is not available and will cause an error. So to make sure our version details are saved to our static record of the work, we will save to a file, then print the file contents back into the notebook. 



```{r rstudio version - not run during knit, eval=FALSE}
# Get RStudio version information
rstudio_info <- RStudio.Version()

# Convert the version information to a string
rstudio_version_str <- paste(
  "RStudio Version Information:\n",
  "Version: ", rstudio_info$version, "\n",
  "Release Name: ", rstudio_info$release_name, "\n",
  "Long Version: ", rstudio_info$long_version, "\n",
  "Mode: ", rstudio_info$mode, "\n",
  "Citation: ", rstudio_info$citation,
  sep = ""
)

# Write the output to a text file
writeLines(rstudio_version_str, "rstudio_version.txt")

```

```{r rstudio version saved from text}

# Read the saved version information from the file
rstudio_version_text <- readLines("rstudio_version.txt")

# Print the version information to the document
rstudio_version_text

```

## Knit workbook to HTML

The last task is to knit the notebook. Our notebook is editable, and can be changed. Deleting code deletes the output, so we could lose valuable details. If we knit the notebook to HTML, we have a permanent static copy of the work.

On the editor pane toolbar, under Preview, select Knit to HTML.

If you have already run Preview, you will see Knit instead of Preview.

The HTML file will be saved in the same directory as the notebook, and with the same filename, but the .Rmd prefix will be replaced by .html. The knit HTML will typically open automatically once complete. If you receive a popup blocker error, click cancel, and in the Files pane of RStudio, single click the gprofiler.html file and select View in Web Browser.

Note that the notebook will only successfully knit if there are no errors in the code. You can 'preview' HTML with code errors.










