---
title: "Non-model species functional enrichment analysis"
output:
  html_document:
    df_print: paged
    self_contained: false
---




```{r  Load libraries, include=FALSE }

library(readr)
library(dplyr)
library(ontologyIndex)
library(tidyverse)
library(clusterProfiler)
library(WebGestaltR)
library(enrichplot)
#library(ggupset)

```


# 0. Working directory

Ensure the 'workshop' directory is your current working directory:

```{r check notebook workdir}
getwd()
```

# 1. Import `emapper` axolotl annotation file, GO ontology file and KEGG Pathways file

## 1.1 Public data sources used for this notebook

- [Reference genome](https://www.axolotl-omics.org/dl/AmexG_v6.0-DD.fa.gz)
- [GTF file](https://www.axolotl-omics.org/dl/AmexT_v47-AmexG_v6.0-DD.gtf.gz)
- [Raw fastq](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA300706)
- [GO 'core' ontology file](https://purl.obolibrary.org/obo/go.obo)


## 1.2 Data processing

The raw reads were processed to counts matrix and DE results broadly following https://github.com/Sydney-Informatics-Hub/RNASeq-DE using the reference genome from www.axolotl-omics.org. Differential gene expression analysis was performed in R with `DESeq2` v 1.46.0, filtering for genes with at least a count of 10 in at least 2 samples. The data comprises 2 groups (proximal blastema and distal blastema) and 2 replicates per group.

A predicted proteome was created by extracting the predicted peptide sequences from the GTF file from www.axolotl-omics.org then filtering for longest isoform per gene with `AGAT` v 1.4.0. The predicated proteome was annotated against GO and KEGG with `eggNOG emapper` v 2.1.12. 

## 1.3 Import annotation files

### 1.3.1 emapper proteome annotation

The `emapper` annotation output contains results against a number of databases including GO and KEGG, which we will focus on today. 

```{r Load eggnog annotation}
eggnog_anno <- read_tsv("AmexG_v6.0-DD.emapper.annotations.txt", show_col_types = FALSE) 
head(eggnog_anno)
```

The raw annotation file provides us with 'term ID to gene ID' mappings for our species. We also need 'term ID to term description' mappings. These files are not organisms specific: we will extract only the terms that are found within our custom species annotation, to make our organism specific version. 

### 1.3.2 GO Core Ontology

For GO, we will use the GO 'core' ontology file, downloaded from https://purl.obolibrary.org/obo/go.obo and included in the data files you downloaded to `workshop` directory earlier. 

We will use the `ontologyIndex` package to retrieve ontology info and save to an object named `ontology` for later use creating the required custom database files for `clusterProfiler` and `WebGestaltR`.  


```{r Extract GO ontology}
ontology <- ontologyIndex::get_ontology(file = "go.obo",
  propagate_relationships = "is_a", #propagates relationships from parent terms to children
  extract_tags = "everything", # retrieve all available details for each term
  merge_equivalent_terms = TRUE) # avoid unecessary redundancy 
```



### 1.3.3 KEGG Pathways

For KEGG, we have both `map` and `ko` IDs in our `emapper` annotation. 

`ko` terms (https://www.genome.jp/kegg/ko.html) represent orthologous groups of genes, which are assigned based on evolutionary relationships and functional similarity, so can provide more precise functional categorisation which can be particularly useful when working with novel species which lack curate dpathway information. 

`map` terms (https://www.genome.jp/kegg/pathway.html) are manually drawn pathway maps representing KEGG database of molecular interaction, reaction and relation networks for: Metabolism, Genetic Information Processing, Environmental Information Processing, Cellular Processes, Organismal Systems, Human Diseases, and Drug Development. 

Today we will be working with the `map` terms due to database download restrictions. 

Free access to the KEGG FTP downloads requires an academic subscription, to which you must confirm to be the  "only user of the KEGG FTP Data". The pathways list was available freely. As a single user, you can request academic access here https://www.pathway.jp/en/academic.html. 

There is an alternate method for using `ko` IDs, which uses the KEGG ontology information available through the `clusterProfiler` functions `enrichKEGG` and `gseKEGG`. An example of the R code can be found here https://github.com/dadrasarmin/enrichment_analysis_for_non_model_organism.  However, this poses a problem: as the novel species gene IDs are assigned to KEGG terms, gene:name duplicate records are identified, and duplicates must be removed in order to avoid errors running the enrichment. This loss of data will have a real impact on the results, with the importance of some terms being underestimated. 

Given these considerations, we will proceed with `map` pathway terms :-) 

The KEGG `map` pathway list was downloaded from https://rest.kegg.jp/list/pathway and downloaded to your `workshop` directory.


```{r import KEGG pathways}
kegg_pathways <- read.table("kegg_pathways_2024-11-13.txt", header = FALSE, sep = "\t", col.names = c("term", "name"))
head(kegg_pathways)
```



 
 

# 2. Import axolotl DE results file and extract gene lists for ORA and GSEA

## 2.1 Import axolotl DE data

Load the DE results file for axolotl: 

```{r load input DE matrix}
de_matrix <- read_tsv("axolotl_DE_results.txt", col_names = TRUE, show_col_types = FALSE)
head(de_matrix)
```


## 2.2 Create the ranked gene list for GSEA

Recall from the last 2 activities that `clusterProfiler` requires a vector object for GSEA, while `WebGestaltR` requires a 2-column dataframe. Since we intend to use both tools, let's create both now: 


```{r create ranked gene vector }

# Create ranked vector for clusterProfiler GSEA
ranked_vector <- setNames(de_matrix$log2FoldChange, de_matrix$geneID) %>% sort(decreasing = TRUE)  # Named vector

# check
head(ranked_vector)
tail(ranked_vector)

```


```{r created ranked gene dataframe}
# Create ranked dataframe for WebGestaltR GSEA# extract ranked dataframe
ranked_df <- de_matrix %>%
  arrange(desc(log2FoldChange)) %>%
  dplyr::select(geneID, log2FoldChange)

# check
head(ranked_df)
tail(ranked_df)
```

## 2.3 Create gene lists for ORA

For ORA, both tools require vector class gene lists. We will filter for adjusted P value < 0.01 and absolute log2 fold change greater than 1.5. The matrix has already filtered out genes with very low counts so we take all genes present as the background. 

```{r create ORA vector gene lists}

# Filter for DEGs and save gene IDs as vector  
degs <- de_matrix %>%
  filter(padj <= 0.01 & abs(log2FoldChange) >= 1.5) %>%
  pull(geneID)  # Extract 

# Extract the background gene list vector 
background <- de_matrix %>%
  pull(geneID)  

# Check number of genes: 
cat("Number of DEGs:", length(degs), "\n")         # Number of DEGs
cat("Number of background genes:", length(background), "\n")   # Number of background genes

# Check format: 
head(degs)
head(background)

```
## 2.4 Save gene lists

Saving any outputs generated from R code is vital to reproducibility! You should include all analysed gene lists within the supplementary materials of your manuscript. 

```{r save gene lists to file}
# Save DEGs
write.table(degs, file = "Axolotl_DEGs.txt", quote = FALSE, col.names = FALSE, row.names = FALSE, sep = "\t")
# Save background
write.table(background, file = "Axolotl_background.txt", quote = FALSE, col.names = FALSE, row.names = FALSE, sep = "\t")
# Save ranked
write.table(ranked_df, file = "Axolotl_rankedFC.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)

```


# 3. Reformat annotation files for `clusterProfiler` GO and KEGG analysis

## 3.1 Create TERM2GENE files

These are 2 column text files with the term ID (one per line) alongside the ID of the gene that maps to the term. A gene can map to many terms and thus be present on multiple lines. A term can be mapped to more than one gene and thus be present on many lines.

Check the column names of the `emapper` annotation file so we know which are the GO and KEGG column names:

```{r colnames anno}
colnames(eggnog_anno)
```
We need `GOs` and `KEGG_Pathway` columns. 

### 3.1.1 GO TERM2GENE

Next, we will extract the GO IDs from the `emapper` annotation file, and wrangle into the correct format for `clusterProfiler` `TERM2GENE`. There are several steps to this - comments have been included to outline what each step is doing.

```{r GO TERM2GENE}
go_term2gene <- eggnog_anno %>%
    dplyr::select(GOs, `#query`) %>% # select the GO column and the query column (axolotl gene ID) 
    dplyr::filter(GOs != "-") %>% # filter out rows where the GO ID is "-" ie no GO annotation for this gene
    separate_rows(GOs, sep = ",") %>% # split comma-delimited list of many GO terms for a gene into separate rows
    dplyr::select(GOs, `#query`) %>% # keep the GO and query columns
    distinct() %>% # remove any duplicate rows 
    drop_na() # remove rows with missing values

# Rename columns to match desired output format
colnames(go_term2gene) <- c("term", "gene")

# Save to file
write.table(go_term2gene, file = "Axolotl_GO_term2gene.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)

# Number of rows: 
cat("Number of GO term2gene rows:", nrow(go_term2gene), "\n")

```

```{r view go term2gene}
# Check first few rows
head(go_term2gene)

```

```{r}

```

### 3.1.2 KEGG TERM2GENE

Here we use the same process as we did above for GO (colum name `GOs`), selecting a different column name for KEGG (`KEGG_Pathway`). 

```{r KEGG TERM2GENE}
kegg_term2gene <- eggnog_anno %>%
    dplyr::select(KEGG_Pathway, `#query`) %>%  # Select the relevant columns
    dplyr::filter(grepl("map", KEGG_Pathway)) %>%  # Keep only rows where KEGG_Pathway contains 'map'
    separate_rows(KEGG_Pathway, sep = ",") %>%  # Split multiple pathways into separate rows
    dplyr::mutate(term = gsub("map:", "", KEGG_Pathway)) %>%  # Remove the "map:" prefix
    dplyr::filter(grepl("^map", term)) %>%  # Filter again to make sure we only have map pathways (after removing "map:")
    dplyr::select(term, `#query`) %>%  # Select the pathway (term) and gene columns
    distinct() %>%  # Remove duplicate rows
    drop_na()  # Remove rows with missing values


# Rename columns to match desired output format
colnames(kegg_term2gene) <- c("term", "gene")

# Save to file
write.table(kegg_term2gene, file = "Axolotl_KEGG-Pathways_term2gene.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)

cat("Number of KEGG term2gene rows:", nrow(kegg_term2gene), "\n")
```


```{r view kegg term2gene}
# View result to check
head(kegg_term2gene)

```

## 3.2 TERM2NAME 

### 3.2.1 GO

The GO Core Ontology file (we have saved in our `workshop` folder and labelled the filename as R object `go_core`) will enable us to assign term descriptions to term IDs and create our `TERM2NAME` files. 


This may take a few moments to run. It will use the `ontology` object we created earlier. 

```{r GO TERM2NAME}

# Create term to name table, removing duplicates, missing values and obsolete terms 
go_term2name <- go_term2gene %>%
    mutate(name = ontology$name[term]) %>%
    dplyr::select(term, name) %>%
    distinct() %>%
    drop_na() %>%
    filter(!grepl("obsolete", name))

# Save to file
write.table(go_term2name, file = "Axolotl_GO_term2name.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)

# Show the first few lines
head(go_term2name)

```



### 3.2.2 KEGG 

The KEGG Pathways file was available for download in the correct format for `TERM2NAME`. 

```{r view kegg pathways file}
head(kegg_pathways)
```

Let's restrict it to include the terms relevant to our analysis, and then print that to a file for reproducibility. 


```{r KEGG TERM2NAME}
kegg_term2name <- kegg_pathways %>%
  dplyr::filter(term %in% kegg_term2gene$term) %>%  # Only keep terms that are in kegg_term2gene
  distinct() %>%  # Remove duplicate entries
  drop_na()  # Remove rows with missing values

# Save the result to a file
write.table(kegg_term2name, file = "Axolotl_KEGG_term2name.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)

# Check first few rows
head(kegg_term2name)
```



## 3.3 Count annotations 

How much of our proteome was annotated? What about our DEGs and background? Genes that do not have any annotation are excluded from enrichment analysis, so having an understanding of the extent of annotation for your novel species is very important when interpreting results! 


Count the number of GO terms found within the genome, and the number of genes with GO annotations: 

```{r report GO annotation counts}

go_total_terms<-nrow(go_term2gene)
print(paste("Total annotations to GO:", go_total_terms))

go_unique_genes <- length(unique(go_term2gene$gene))
print(paste("Number of unique genes with 1 or more annotation terms:", go_unique_genes))

```

And for KEGG:

```{r report KEGG annotation counts}

kegg_total_terms<-nrow(kegg_term2gene)
print(paste("Total annotations to KEGG Pathways:", kegg_total_terms))

kegg_unique_genes <- length(unique(kegg_term2gene$gene))
print(paste("Number of unique genes with 1 or more annotation terms:", kegg_unique_genes))

```

47,196 putative axolotl proteins were annotated. That's around 1/4 of our genes mapped to KEGG Pathways, and less than half of our genes mapped to GO! Ouch. As much as we expect this with uncurated novel species genomes, it's still unpleasant to face :-)


What of the genes in our gene list specifically? We have an uncurated proteome, yet the genes in our input matrix were expressed at a meaningful level within axolotl, so these may actually have a higher annotation percentage than all genes in the proteome. 


```{r report gene list GO annotation percentages}
# Filter the term2gene table to only include genes in the background gene list
go_filtered_term2gene <- go_term2gene %>% filter(gene %in% background)

# Count the number of unique background genes with at least one GO term
unique_genes_with_go <- go_filtered_term2gene %>% distinct(gene) %>% nrow()

# Calculate the percentage of background genes that have GO annotations
percent_go_unique <- (unique_genes_with_go / length(background)) * 100

# Print results
cat("Number of input genes with GO annotations:", unique_genes_with_go, "(",percent_go_unique,"%)\n")


```

```{r report gene list KEGG annotation percentages}
# Filter the term2gene table to only include genes in the background gene list
kegg_filtered_term2gene <- kegg_term2gene %>% filter(gene %in% background)

# Count the number of unique background genes with at least one GO term
unique_genes_with_kegg <- kegg_filtered_term2gene %>% distinct(gene) %>% nrow()

# Calculate the percentage of background genes that have GO annotations
percent_kegg_unique <- (unique_genes_with_kegg / length(background)) * 100

# Print results
cat("Number of input genes with KEGG Pathways annotations:", unique_genes_with_kegg, "(",percent_kegg_unique,"%)\n")


```

As expected, the annotation % is higher for expressed genes than all predicted genes, and very much higher than the GTF of 99,088 predicted gene models (!!!) with an annotation rate of 21.6%.   

This highlights a major caveat when performing FEA on non-model species: the results are only as good as the annotations behind them. Therefore, all results must be interpreted with caution. For many novel (and under-funded) species, there are little opportunities (at present) to improve the annotation. Some in-silico predicted genes appear to be highly expressed and significantly regulated yet have no significant similarity to anything in the non-redundant nucleotide or protein databases. When working with datasets like this, it is critical to explore those individual genes through other methods, in addition to trying to garner some higher level overview such as we aim to obtain from FEA. Hopefully, recent advances in AI protein modelling can help provide insights into the functions of these novel genes. 

For the axolotl with only 22% of predicted genes annotated, its clear that the in-silico gene predictions within the GTF file require much curation! 

# 4. Run `clusterProfiler` universal FEA functions `enricher` and `GSEA`

In the interest of time, and to try and cover as many options as possible, let's do ORA with GO and GSEA with KEGG for both tools. 

## 4.1 clusterProfiler ORA of GO terms

The `enricher` function is the 'universal' ORA option that accepts the `TERM2GENE` and `TERM2NAME` files we have just created. 

Let's review the help page:

```{r help enricher}
?clusterProfiler::enricher
```

There are parameters for both adjusted P value and q value. Terms must pass all thresholds (unadjusted P, adjusted P, and q value) so the important filter will be the most stringent test applied. Let's go with BH and 0.05 which we have used regularly within this workshop and are fairly common choices in the field. 


```{r CP GO ORA }

cp_go_ora <- enricher(
  gene = degs,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  universe = background,
  minGSSize = 10,
  maxGSSize = 500,
  TERM2GENE = go_term2gene,
  TERM2NAME = go_term2name
)
```


```{r view CP GP ORA result summary}
cp_go_ora

```

91 significantly enriched terms at P.adj < 0.05. 

Note that we used the gene list object `degs` which had 247 genes, but the tool has applied the input size as 145 - this is because it is automatically discarding any that do not have annotations. Results would be the same if we instead used `annotated_degs` object. Likewise, the background size is being reported as 15072 (the number annotated) not 24,419 (the total in background list). 


Save the results to a text file:


```{r save CP GO ORA table}
file <- "Axolotl_clusterProfiler_GO_ORA_results.tsv"
write.table(cp_go_ora, file, sep = "\t", quote = FALSE, row.names = FALSE)	

```


Let's visualise with one of my favourite `enrichplot` plots, the treeplot! Another advantage of this plot is that it can be used for both ORA and GSEA results, so we can compare more easily. We will add a custom subtitle that informs the number of DEGs that were actually annotated and included in the FEA, so anyone reviewing the plot will understand that caution must be exercised when interpreting the results. 

```{r treeplot CP GO ORA, , fig.width=12}
# calculate pairwise similarities between the enriched terms
cp_go_ora <- enrichplot::pairwise_termsim(cp_go_ora)
p<- enrichplot::treeplot(cp_go_ora, 
  showCategory = 15, 
  color = "p.adjust", 
  cluster.params = list(label_words_n = 5)
)

# Add annotations (number of input genes and number of input genes with GO terms)
num_genes <- length(degs)
genes_with_GO_terms <- sum(degs %in% go_term2gene$gene)

# Print the plot with custom sub-title
p <- p + ggtitle("clusterProfiler ORA of GO terms") + labs(subtitle = paste("Input genes:", num_genes, "| Input genes with GO terms:", genes_with_GO_terms))
print(p)
```


There's a lot of skin and muscle stuff, which we expect to be expressed in the blastema. As for why they are dysregulated? This is a dummy experiment from public RNAseq, with poor replication, and may not even be the right experiment type for this question, so let's not hope for too many clear answers :-) 

## 4.2 clusterProfiler GSEA of KEGG terms

The `GSEA` function is the 'universal' GSEA option that accepts the `TERM2GENE` and `TERM2NAME` files we have just created. 

Let's review the help page:

```{r help GSEA}
?clusterProfiler::GSEA
```

Recall from our `clusterProfiler` session with human data that we needed to add `nPermSimple = 10000` to avoid a warning about "unbalanced (positive and negative) gene-level statistic value" and reduce `eps` to zero to avoid a warning about obtaining better P value estimates . Let's do this from the start.

```{r CP KEGG GSEA}
cp_kegg_gsea <- GSEA(
  geneList = ranked_vector, 
  exponent = 1, 
  minGSSize = 10, 
  maxGSSize = 500, 
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  TERM2GENE = kegg_term2gene, 
  TERM2NAME = kegg_term2name, 
  seed = 123, 
  by = "fgsea",
  nPermSimple = 10000
)

```


```{r view CP KEGG GSEA result summary}
cp_kegg_gsea
```

29 enriched terms. 

Let's treeplot!

```{r treeplot CP KEGG GSEA, fig.width=12}
# calculate pairwise similarities between the enriched terms
cp_kegg_gsea <- enrichplot::pairwise_termsim(cp_kegg_gsea)
p<- enrichplot::treeplot(cp_kegg_gsea, 
  showCategory = 15, 
  color = "p.adjust", 
  cluster.params = list(label_words_n = 5)
)

# Add annotations (number of input genes and number of input genes with GO terms)
# Use background since all genes for ranked are in background
num_genes <- length(background)
genes_with_kegg_terms <- sum(background %in% kegg_term2gene$gene)

# Print the plot with custom sub-title
p <- p + ggtitle("clusterProfiler GSEA of KEGG Pathways") + labs(subtitle = paste("Input genes:", num_genes, "| Input genes with KEGG pathway terms:", genes_with_kegg_terms))
print(p)
```

Some muscle stuff, some cull junction stuff, and some infection-related terms. This can be common in FEA, many genes involved in infection responses are also part of broader stress response pathways. These genes may be activated under different conditions, such as environmental stress, tissue injury, or other disruptions to homeostasis, which are common in various types of experiments. Pathways related to immune responses can also be interconnected with pathways controlling inflammation, wound healing, and metabolic processes. As a result, infection-related pathways can appear in enrichment analysis even when the experimental conditions don't directly involve infection. This does not mean the result is spurious - it just requires that you exercise pragmatism, employ a basic understanding of the statistical approach, and commit to interpreting the results in the context of your experiment. Remember that the FEA results are to bring a large list of genes down to a high level overview to help guide further investigation rather than give a clear answer to your experiment. 


I favour a volcano plot for GSEA, so we can see positive vs negative NES. This is part of `ggplot`, not `enrichplot`, where the `volplot` is only for ORA. 


```{r volcano plot CP gsea kegg, ,  fig.height=6, fig.width=9}

p<- ggplot(cp_kegg_gsea@result, aes(x = enrichmentScore, y = -log10(p.adjust), color = p.adjust)) + 
  geom_point(alpha = 0.7, size = 2) +  # Adjust point size
  scale_color_gradient(low = "blue", high = "red") +  # Color by p.adjust values
  theme_minimal() + 
  labs(title = "clusterProfiler GSEA of KEGG Pathways", 
       x = "Enrichment Score (NES)", 
       y = "-log10(Adjusted P-value)",
       color = "Adjusted P-value") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels for readability
  geom_vline(xintercept = 0, linetype = "dashed", color = "black") +  # Add vertical line at x=0
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +  # Add horizontal line at p=0.05 cutoff
  geom_text(aes(label = Description), 
            hjust = 0.5, 
            vjust = -0.5,  # Move labels higher off the points
            size = 3, 
            check_overlap = TRUE, 
            alpha = 0.7)  # Add labels for each pathway term

print(p)
```

Interesting that all terms except one have leading edge genes that are upregulated in distal compared to proximal (the reference level)!


# 5. Reformat annotation files for `WebGestaltR` GO and KEGG analysis

GMT files must have `.gmt` suffix and description files must have `.des` suffix. 

## 5.1 Create GMT objects

The GMT files need links for all of the terms, so that we can have that handy link-out to enriched terms from the HTML report we experienced in the last activity. This is actually pretty simple to do thanks to consistent URLs. 

For GO, we just need to paste the term ID to the end of this link https://www.ebi.ac.uk/QuickGO/term/<GOID>

And for KEGG, we need to paste the map ID to the end of this link: https://www.genome.jp/dbget-bin/www_bget?<mapID>


### 5.1.1 GO GMT

Note that in the below code, the first command is identical the one that created the `go_term2gene` object earlier in the notebook. we could just use the `go_term2gene` object and skip step of the below code, using `go_term2gene` as input for step 2 rather than `go_data`. This code duplication is intentional, so that this code chunk is standalone for re-use and re-purpose. 

```{r GO GMT}

# Step 1: Extract relevant columns (GO terms and gene IDs) from eggnog_anno
go_data <- eggnog_anno %>%
  dplyr::select(GOs, `#query`) %>%  # Select the GO terms and the gene IDs
  dplyr::filter(GOs != "-") %>%  # Filter out rows where the GO ID is missing ("-")
  separate_rows(GOs, sep = ",") %>%  # Split comma-delimited list of GO terms into separate rows
  dplyr::select(GOs, `#query`) %>%  # Keep GO terms and gene IDs columns
  distinct() %>%  # Remove duplicates
  drop_na()  # Drop any rows with missing values

# Rename columns to match the format (term, gene)
colnames(go_data) <- c("term", "gene")

# Step 2: Create external links for each GO term (link to QuickGO)
go_data <- go_data %>%
  dplyr::mutate(external_link = paste0("https://www.ebi.ac.uk/QuickGO/term/", term))

# Step 3: Group genes by GO term and concatenate gene list by tab
go_term_grouped <- go_data %>%
  dplyr::group_by(term) %>%
  dplyr::summarize(genes = paste(gene, collapse = "\t"), .groups = "drop")

# Step 4: Add the external link for each GO term
go_term_grouped <- go_term_grouped %>%
  dplyr::left_join(go_data %>% dplyr::select(term, external_link) %>% distinct(), by = "term")

# Step 5: Create the final GMT format entry (term ID, external link, and gene list)
go_gmt <- go_term_grouped %>%
  dplyr::mutate(gmt_entry = paste(term, external_link, genes, sep = "\t")) %>%
  dplyr::select(gmt_entry)

# Save to file
write.table(go_gmt, file = "Axolotl_GO.gmt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```




### 5.1.2 KEGG GMT

As above, for clarity we have avoided using the `TERM2GENE` object to ensure this code chunk can be standalone.  


```{r KEGG GMT}
# Step 1: Extract relevant columns (KEGG Pathway and gene IDs) from eggnog_anno
kegg_data <- eggnog_anno %>%
  dplyr::select(KEGG_Pathway, `#query`) %>%  # Select the KEGG Pathway and gene ID columns
  dplyr::filter(grepl("map", KEGG_Pathway)) %>%  # Keep only rows where KEGG_Pathway contains 'map'
  separate_rows(KEGG_Pathway, sep = ",") %>%  # Split multiple pathways into separate rows
  dplyr::mutate(term = gsub("map:", "", KEGG_Pathway)) %>%  # Remove the "map:" prefix
  dplyr::filter(grepl("^map", term)) %>%  # Filter again to keep only 'map' pathways (after removing "map:")
  dplyr::select(term, `#query`) %>%  # Select the KEGG Pathway and gene ID columns
  distinct() %>%  # Remove duplicate rows
  drop_na()  # Remove rows with missing values

# Ensure the column is properly named
colnames(kegg_data)[colnames(kegg_data) == "#query"] <- "gene"

# Step 2: Create external links for each KEGG pathway
kegg_data <- kegg_data %>%
  dplyr::mutate(external_link = paste0("https://www.genome.jp/dbget-bin/www_bget?", term))

# Step 3: Group by KEGG pathway term and concatenate the gene list
kegg_term_grouped <- kegg_data %>%
  dplyr::group_by(term) %>%
  dplyr::summarize(genes = paste(gene, collapse = "\t"), .groups = "drop")

# Step 4: Add the external link for each KEGG pathway
kegg_term_grouped <- kegg_term_grouped %>%
  dplyr::left_join(kegg_data %>% dplyr::select(term, external_link) %>% distinct(), by = "term")

# Step 5: Create the final GMT format entry (Pathway, External Link, Genes)
kegg_gmt <- kegg_term_grouped %>%
  dplyr::mutate(gmt_entry = paste(term, external_link, genes, sep = "\t")) %>%
  dplyr::select(gmt_entry)

# Save to file
write.table(kegg_gmt, file = "Axolotl_KEGG-pathways.gmt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

```

## 5.2 Create description objects

### 5.2.1 GO description

The description file is identical to `clusterProfiler` `TERM2NAME` . Again, we don't want to use any code from the creation of `clusterProfiler` objects to ensure this part can be used alone.  


```{r GO description }

# Step 1: Get terms from ontology object created from go.obo with ontologyIndex function: 
ontology_term_names <- ontology$name 

# Step 2: Filter and separate GO terms from the annotation file
# We filter out rows where no GO terms are assigned and separate comma-delimited GO terms
go_terms <- eggnog_anno %>%
  dplyr::select(GOs, `#query`) %>%
  dplyr::filter(GOs != "-") %>%  # Keep only rows with GO terms
  separate_rows(GOs, sep = ",") %>%
  dplyr::mutate(term = GOs) %>%
  dplyr::select(term, `#query`) %>%
  distinct() %>%
  drop_na()  # Drop rows with missing values

# Step 3: Create the description by matching GO terms to their names in the ontology
go_des <- go_terms %>%
  dplyr::mutate(name = ontology_term_names[term]) %>%  # Map term to its name from the ontology
  dplyr::select(term, name) %>%  # Keep only the term and name
  distinct() %>%  # Remove duplicates
  drop_na() %>%  # Remove rows with missing values
  filter(!grepl("obsolete", name))  # Remove obsolete terms if present

# Save to file
write.table(go_des, file = "Axolotl_GO.des", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

# Check
head(go_des)
```



### 5.2.2 KEGG description file 

More extra work for the sake of portability :-)

```{r KEGG description}
# Get columns from kegg gmt object 
kegg_gmt_columns <- kegg_gmt %>%
  separate(gmt_entry, into = c("term", "external_link", "genes"), sep = "\t")

# Create the kegg_des table by joining the pathways file with the species-specific terms from kegg_gmt
kegg_des <- kegg_pathways %>%
  dplyr::filter(term %in% kegg_gmt_columns$term) %>%
  dplyr::select(term, name) %>%
  distinct() %>%
  drop_na()

# Save to file
write.table(kegg_des, file = "Axolotl_KEGG-pathways.des", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

# Check the first few rows
head(kegg_des)
```




# 6. Run `WebGestaltR` ORA and GSEA with custom database files

We will use the same approach as with `clusterProfiler` and run ORA with GO and GSEA with KEGG for both tools. 

## 6.1 WebGestaltR ORA of GO terms


```{r WGR ORA GO }

outputDirectory <- "WebGestaltR_results" 
project <- "Axolotl_ORA_GO"

WebGestaltR(
    organism = "others",                          # must specify 'others' when using custom db files
    enrichMethod = "ORA",                         # Perform ORA, GSEA or NTA
    interestGene = degs,                          # gene list of interest
    referenceGene = background,                   # background genes
    enrichDatabaseFile = "Axolotl_GO.gmt",        # the custom gmt file
    enrichDatabaseDescriptionFile = "Axolotl_GO.des",   # the custom description file  
    isOutput = TRUE,                              # Set to FALSE if you don't want files saved to disk
    fdrMethod = "BH",                             # Benjamini-Hochberg multiple testing correction
    sigMethod = "fdr",                            # Significance method ('fdr' or 'top')
    fdrThr = 0.05,                                # FDR significance threshold
    minNum = 10,                                   # Minimum number of genes per category
    maxNum = 500,                                 # Maximum number of genes per category
    boxplot = TRUE,
    outputDirectory = outputDirectory,
    projectName = project
)

```


Open the HTML report `WebGestaltR_results/Project_Axolotl_ORA_GO/Report_Axolotl_ORA_GO.html` from the files pane in a browser. 

Note some term similarity to what we have seen with the past 2 analyses (that's reassuring!) 

Change the 'Enrichment Results' view from table to 'Bar  chart', then try the 'Affinity propagation' and 'Weighted set cover' term clustering algorithms. Both methods help focus the results. 

Confirm that our GMT file correctly included the term link by selecting a term and clicking the hyperlink at `Analyte set`. Pretty neat huh :-) 

## 6.2 WebGestaltR GSEA of KEGG Pathways 

This will take slightly longer than ORA. We will set threads to 7 to speed it up as much as we can.  

There is no `seed` parameter for `WebGestaltR` GSEA as there is for `clusterProfiler`. We can set it in R instead with `set.seed()`. 

```{r set seed}
set.seed(123)
```


```{r WGR GSEA KEGG }

outputDirectory <- "WebGestaltR_results" 
project <- "Axolotl_GSEA_KEGG"

suppressWarnings({ WebGestaltR(
    organism = "others",                          # must specify 'others' when using custom db files
    enrichMethod = "GSEA",                        # Perform ORA, GSEA or NTA
    interestGene = ranked_df,                     # ranked dataframe
    enrichDatabaseFile = "Axolotl_KEGG-pathways.gmt",        # the custom gmt file
    enrichDatabaseDescriptionFile = "Axolotl_KEGG-pathways.des",   # the custom description file  
    isOutput = TRUE,                              # Set to FALSE if you don't want files saved to disk
    fdrMethod = "BH",                             # Benjamini-Hochberg multiple testing correction
    sigMethod = "fdr",                            # Significance method ('fdr' or 'top')
    fdrThr = 0.05,                                # FDR significance threshold
    minNum = 10,                                   # Minimum number of genes per category
    maxNum = 500,                                 # Maximum number of genes per category
    outputDirectory = outputDirectory,
    projectName = project,
    nThreads = 7
) })

```

Open the HTML report `WebGestaltR_results/Project_Axolotl_GSEA_KEGG/Report_Axolotl_GSEA_KEGG.html` from the files pane in a browser.

Expand 'Job summary' to read that "22 positive related categories and no negative related categories" are significant in this analysis. This is in contrast to the one negative category we observed when running KEGG GSEA with `clusterProfiler`. We expect some differences between these tools.

Compare the tabular results in this report to the treeplot we produced under code chunk `treeplot CP KEGG GSEA`. There are a lot of shared terms, and this is reassuring. 

# 7. Save versions and session details

## GO database version 

Print the database version of GO Core Ontology used:

```{r print GO core ontology file version}
# Read go.obo lines
lines <- readLines("go.obo")

# Use grep to pull "data-version"
version <- grep("data-version", lines, value = TRUE)

# Print version
cat("GO version from go.obo file:", version, "\n")

```

## KEGG Pathways database version 

The KEGG pathways file does not contain any version details within the file contents, but does have the date saved in the name of the file that was imported into this workbook. Adding the date of download was done manually, and is always recommended practice for files and databases that do not contain any date or version details. 


## R version and R package versions

```{r info }
sessionInfo()
```


## RStudio version

And RStudio version. Typically, we would simply run `RStudio.Version()` to print the version details. However, when we knit this document to HTML, the `RStudio.Version()` function is not available and will cause an error. So to make sure our version details are saved to our static record of the work, we will save to a file, then print the file contents back into the notebook. 


```{r rstudio version - not run during knit, eval=FALSE}
# Get RStudio version information
rstudio_info <- RStudio.Version()

# Convert the version information to a string
rstudio_version_str <- paste(
  "RStudio Version Information:\n",
  "Version: ", rstudio_info$version, "\n",
  "Release Name: ", rstudio_info$release_name, "\n",
  "Long Version: ", rstudio_info$long_version, "\n",
  "Mode: ", rstudio_info$mode, "\n",
  "Citation: ", rstudio_info$citation,
  sep = ""
)

# Write the output to a text file
writeLines(rstudio_version_str, "rstudio_version.txt")

```

```{r rstudio version saved from text}

# Read the saved version information from the file
rstudio_version_text <- readLines("rstudio_version.txt")

# Print the version information to the document
rstudio_version_text

```



# 8.  Knit workbook to HTML

The last task is to knit the notebook. Our notebook is editable, and can be changed. Deleting code deletes the output, so we could lose valuable details. If we knit the notebook to HTML, we have a permanent static copy of the work.

On the editor pane toolbar, under Preview, select Knit to HTML.

If you have already run Preview, you will see Knit instead of Preview.

The HTML file will be saved in the same directory as the notebook, and with the same filename, but the .Rmd prefix will be replaced by .html. The knit HTML will typically open automatically once complete. If you receive a popup blocker error, click cancel, and in the Files pane of RStudio, single click the gprofiler.html file and select View in Web Browser.

Note that the notebook will only successfully knit if there are no errors in the code. You can 'preview' HTML with code errors.






